From d17339500580aa8914ca57778027565ab1228062 Mon Sep 17 00:00:00 2001
From: "mujiburrahimank@maxlinear.com"
 <mujiburrahimank@lvdocker7.lvsw.maxlinear.com>
Date: Fri, 28 Nov 2025 05:47:40 -0800
Subject: [PATCH] WLANRTSYS-97395 : Upstep hostap-ng to the latest opensource
 commit

Fixing Compilation issues after upstep to opensource commit
95ad71157e3095c52d5212855ca232832b1b5085
---
 afc/afcd.c                         |  2 ++
 hostapd/hostapd_cli.c              | 10 ------
 src/ap/acs.c                       |  2 +-
 src/ap/ap_config.c                 |  3 +-
 src/ap/ap_drv_ops.c                | 37 ++++++++++++++++++++
 src/ap/ap_mlme.c                   |  1 +
 src/ap/beacon.c                    | 31 +++++++++++------
 src/ap/dfs.c                       | 19 +++++++++++
 src/ap/drv_callbacks.c             |  6 ++--
 src/ap/hostapd.c                   | 23 ++++++++++++-
 src/ap/hostapd.h                   | 25 +-------------
 src/ap/ieee802_11.c                | 46 ++++++-------------------
 src/ap/ieee802_11_eht.c            |  2 +-
 src/ap/ieee802_11_ht.c             | 49 --------------------------
 src/ap/ieee802_11_shared.c         |  5 +--
 src/ap/mxl_beacon.c                |  2 +-
 src/ap/mxl_drv_callbacks.c         |  1 +
 src/ap/mxl_hostapd.c               |  6 ++--
 src/ap/mxl_ieee802_11.c            |  2 +-
 src/ap/mxl_mld.c                   | 10 +++---
 src/ap/mxl_mld.h                   |  2 ++
 src/ap/mxl_utils.c                 |  2 +-
 src/ap/mxl_wpa_auth.c              | 11 +++---
 src/ap/pmksa_cache_auth.c          |  1 +
 src/ap/rrm.c                       |  2 ++
 src/ap/sta_info.c                  | 45 ++++++++++++------------
 src/ap/sta_info.h                  |  5 +++
 src/ap/utils.c                     |  2 +-
 src/ap/wds_wpa.c                   |  2 +-
 src/ap/wnm_ap.c                    |  2 +-
 src/ap/wpa_auth.c                  | 39 +++++++++++----------
 src/ap/wpa_auth.h                  |  6 ++++
 src/ap/wpa_auth_glue.c             | 12 +++----
 src/ap/wpa_auth_glue.h             |  3 +-
 src/ap/wpa_auth_ie.c               |  5 ---
 src/common/defs.h                  | 14 ++++++++
 src/common/hw_features_common.c    | 20 +++++++++++
 src/common/hw_features_common.h    | 14 ++++++++
 src/common/ieee802_11_common.c     |  2 ++
 src/common/wpa_common.c            |  2 +-
 src/drivers/driver.h               |  2 ++
 src/drivers/driver_common.c        |  3 +-
 src/drivers/driver_nl80211.c       | 55 +++++++++++++++++++++++++++---
 src/drivers/driver_nl80211_capa.c  |  2 +-
 src/drivers/driver_nl80211_event.c |  5 +--
 src/drivers/mxl_driver_nl80211.c   |  6 ++--
 wpa_supplicant/Makefile            |  1 -
 wpa_supplicant/bgscan_simple.c     |  2 ++
 wpa_supplicant/dbus/dbus_new.c     |  1 +
 wpa_supplicant/driver_i.h          |  2 +-
 wpa_supplicant/mesh.c              | 22 +++++++++++-
 wpa_supplicant/mxl_supplicant.c    |  3 +-
 wpa_supplicant/p2p_supplicant.c    | 14 ++++++++
 wpa_supplicant/wpa_supplicant.c    |  4 +++
 54 files changed, 367 insertions(+), 228 deletions(-)

diff --git a/afc/afcd.c b/afc/afcd.c
index 8c173dab5..b103988c7 100644
--- a/afc/afcd.c
+++ b/afc/afcd.c
@@ -120,7 +120,9 @@ static int afcd_send_request(struct curl_ctx *ctx, unsigned char *request)
 
 out_slist_free_all:
 	curl_slist_free_all(headers);
+#ifndef CONFIG_VENDOR_MXL
 out_easy_cleanup:
+#endif /* CONFIG_VENDOR_MXL */
 	curl_easy_cleanup(curl);
 out_global_cleanup:
 	curl_global_cleanup();
diff --git a/hostapd/hostapd_cli.c b/hostapd/hostapd_cli.c
index c008f6df1..2cbcfe47d 100644
--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -1445,14 +1445,6 @@ static int hostapd_cli_cmd_disable(struct wpa_ctrl *ctrl, int argc,
 	return wpa_ctrl_command(ctrl, "DISABLE");
 }
 
-#ifdef CONFIG_VENDOR_MXL
-static int hostapd_cli_cmd_stop_ap(struct wpa_ctrl *ctrl, int argc,
-				      char *argv[])
-{
-	return wpa_ctrl_command(ctrl, "STOP_AP");
-}
-#endif /* CONFIG_VENDOR_MXL */
-
 static int hostapd_cli_cmd_enable_mld(struct wpa_ctrl *ctrl, int argc,
 				      char *argv[])
 {
@@ -3046,8 +3038,6 @@ static const struct hostapd_cli_cmd hostapd_cli_commands[] = {
 	  "<MAC addr> = get HE PHY Channel Width Set field for specified STA" },
 	{ "get_sta_he_caps", hostapd_cli_cmd_get_sta_he_caps, NULL,
 	  "<MAC addr> = get STA HE capabilities" },
-	{ "stop_ap", hostapd_cli_cmd_stop_ap, NULL,
-	  "= stop hostapd AP on current interface" },
 #endif /* CONFIG_VENDOR_MXL */
 #ifdef CONFIG_WIFI_CERTIFICATION
 	{ "pmk", hostapd_cli_cmd_pmk, NULL,
diff --git a/src/ap/acs.c b/src/ap/acs.c
index 19d4d3650..6dbba8f46 100644
--- a/src/ap/acs.c
+++ b/src/ap/acs.c
@@ -1541,7 +1541,7 @@ int mxl_acs_find_center_chan(int freq, int width) {
 		case 40:	return acs_get_bw_center_chan(freq, ACS_BW40);
 		case 80:	return acs_get_bw_center_chan(freq, ACS_BW80);
 		case 160:	return acs_get_bw_center_chan(freq, ACS_BW160);
-		case 320:	return acs_get_bw_center_chan(freq, ACS_BW320);
+		case 320:	return acs_get_bw_center_chan(freq, ACS_BW320_1);
 		default:
 			wpa_printf(MSG_ERROR,
 				"TPE: unsupported width for EIRP channel search");
diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index ffa9c3b09..1feb48fb9 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -19,8 +19,9 @@
 #include "common/sae.h"
 #include "eap_common/eap_wsc_common.h"
 #include "eap_server/eap.h"
-#include "wpa_auth.h"
+#include "hostapd.h"
 #include "sta_info.h"
+#include "wpa_auth.h"
 #include "airtime_policy.h"
 #include "ap_config.h"
 #ifdef CONFIG_VENDOR_MXL
diff --git a/src/ap/ap_drv_ops.c b/src/ap/ap_drv_ops.c
index 8d5136140..37130a86a 100644
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -547,8 +547,10 @@ int hostapd_sta_add(struct hostapd_data *hapd,
 		params.mld_link_addr = link_addr;
 		params.mld_link_sta = mld_link_sta;
 		/* Copy EML capabilities of ML STA */
+#ifndef CONFIG_VENDOR_MXL
 		if (link_addr)
 			params.eml_cap = eml_cap;
+#endif /* CONFIG_VENDOR_MXL */
 	}
 #endif /* CONFIG_IEEE80211BE */
 
@@ -695,6 +697,7 @@ int hostapd_set_freq(struct hostapd_data *hapd, enum hostapd_hw_mode mode,
 	struct hostapd_freq_params data;
 	struct hostapd_hw_modes *cmode = hapd->iface->current_mode;
 
+#ifndef CONFIG_VENDOR_MXL
 	if (hostapd_set_freq_params(&data, mode, freq, channel, edmg,
 				    edmg_channel, ht_enabled,
 				    vht_enabled, he_enabled, eht_enabled,
@@ -707,6 +710,28 @@ int hostapd_set_freq(struct hostapd_data *hapd, enum hostapd_hw_mode mode,
 				    &cmode->eht_capab[IEEE80211_MODE_AP] :
 				    NULL, hostapd_get_punct_bitmap(hapd)))
 		return -1;
+#else /* CONFIG_VENDOR_MXL */
+if (hostapd_set_freq_params(&data, mode, freq, channel, edmg,
+                edmg_channel, ht_enabled,
+                vht_enabled, he_enabled, eht_enabled,
+                sec_channel_offset, oper_chwidth,
+                center_segment0, center_segment1,
+                cmode ? cmode->vht_capab : 0,
+                cmode ?
+                &cmode->he_capab[IEEE80211_MODE_AP] : NULL,
+                cmode ?
+                &cmode->eht_capab[IEEE80211_MODE_AP] :
+                NULL)) {
+		wpa_printf(MSG_ERROR, "hostapd_set_freq_params failed: mode=%i, freq=%i, "
+				"channel=%i, ht_enabled=%i, vht_enabled=%i,  he_enabled %i, "
+				"sec_channel_offset=%i, oper_chcwidth=%i, "
+				"center_segment0=%i, center_segment1=%i",
+				mode, freq, channel, ht_enabled, vht_enabled, he_enabled,
+				sec_channel_offset, oper_chwidth, center_segment0,
+				center_segment1);
+		return -1;
+	}
+#endif /* CONFIG_VENDOR_MXL */
 
 	if (hapd->driver == NULL)
 		return 0;
@@ -1118,6 +1143,7 @@ int hostapd_start_dfs_cac(struct hostapd_iface *iface,
 		return -1;
 	}
 
+#ifndef CONFIG_VENDOR_MXL
 	if (hostapd_set_freq_params(&data, mode, freq, channel, 0, 0,
 				    ht_enabled,
 				    vht_enabled, he_enabled, eht_enabled,
@@ -1128,6 +1154,17 @@ int hostapd_start_dfs_cac(struct hostapd_iface *iface,
 				    &cmode->he_capab[IEEE80211_MODE_AP],
 				    &cmode->eht_capab[IEEE80211_MODE_AP],
 				    hostapd_get_punct_bitmap(hapd))) {
+#else /* CONFIG_VENDOR_MXL */
+	if (hostapd_set_freq_params(&data, mode, freq, channel, 0, 0,
+				    ht_enabled,
+				    vht_enabled, he_enabled, eht_enabled,
+				    sec_channel_offset,
+				    oper_chwidth, center_segment0,
+				    center_segment1,
+				    cmode->vht_capab,
+				    &cmode->he_capab[IEEE80211_MODE_AP],
+				    &cmode->eht_capab[IEEE80211_MODE_AP])) {
+#endif /* CONFIG_VENDOR_MXL */
 		wpa_printf(MSG_ERROR, "Can't set freq params");
 		return -1;
 	}
diff --git a/src/ap/ap_mlme.c b/src/ap/ap_mlme.c
index 309e69a3f..3589282be 100644
--- a/src/ap/ap_mlme.c
+++ b/src/ap/ap_mlme.c
@@ -14,6 +14,7 @@
 #include "common/ieee802_11_defs.h"
 #include "ieee802_11.h"
 #include "wpa_auth.h"
+#include "hostapd.h"
 #include "sta_info.h"
 #include "ap_mlme.h"
 #include "hostapd.h"
diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index 7bb787e95..4f448ba5b 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -913,7 +913,9 @@ static u8 * hostapd_probe_resp_fill_elems(struct hostapd_data *hapd,
 		}
 	}
 
+#if 0
 	if (!params->is_ml_sta_info) {
+#endif
 		*pos++ = WLAN_EID_SSID;
 		/* For multibss case, if the transmitted VAP is hidden, it must not reveal its own ssid
 		 * while transmiting probe resp frame on behalf of a non-transmited vap
@@ -933,7 +935,7 @@ static u8 * hostapd_probe_resp_fill_elems(struct hostapd_data *hapd,
 			os_memcpy(pos, hapd->conf->ssid.ssid, hapd->conf->ssid.ssid_len);
 			pos += hapd->conf->ssid.ssid_len;
 		}
-	}
+//	}
 #endif /* CONFIG_VENDOR_MXL */
 
 	/* Supported rates */
@@ -1997,8 +1999,6 @@ void handle_probe_req(struct hostapd_data *hapd,
 	params.is_p2p = !!elems.p2p;
 	params.known_bss = elems.mbssid_known_bss;
 	params.known_bss_len = elems.mbssid_known_bss_len;
-#ifdef CONFIG_VENDOR_MXL
-	params.is_ml_sta_info = false;
 #ifdef CONFIG_VENDOR_MXL
 	params.is_he_in_req = (elems.he_capabilities != NULL);
 	params.probe_req_ssid = elems.ssid;
@@ -2159,17 +2159,14 @@ u8 * hostapd_unsol_bcast_probe_resp(struct hostapd_data *hapd,
 	if (!is_6ghz_op_class(hapd->iconf->op_class))
 		return NULL;
 
-#ifndef CONFIG_VENDOR_MXL
 	ubpr->unsol_bcast_probe_resp_interval =
 		hapd->conf->unsol_bcast_probe_resp_interval;
-#else CONFIG_VENDOR_MXL
-	params->unsol_bcast_probe_resp_interval =
-		hapd->conf->unsol_bcast_probe_resp_interval;
-#endif /* CONFIG_VENDOR_MXL */
 
-	if (!params->unsol_bcast_probe_resp_interval) {
+#ifdef CONFIG_VENDOR_MXL
+	if (!ubpr->unsol_bcast_probe_resp_interval) {
 		return NULL;
 	}
+#endif /* CONFIG_VENDOR_MXL */
 	os_memset(&probe_params, 0, sizeof(probe_params));
 	probe_params.req = NULL;
 	probe_params.is_p2p = false;
@@ -2635,7 +2632,6 @@ int ieee802_11_build_ap_params(struct hostapd_data *hapd,
 #ifdef CONFIG_SCS_MXL
 	tail_len += mxl_hostapd_eid_wce_len(hapd);
 #endif /* CONFIG_SCS_MXL */
-#endif /* CONFIG_VENDOR_MXL */
 
 	if (hapd->iconf->multibss_enable && (MULTIBSS_REFERENCE_BSS_IDX == hapd->conf->bss_idx)) {
 		multiple_bssid_ie_len = multiple_bssid_ie_get_len(hapd, true, NULL, 0);
@@ -3247,6 +3243,7 @@ static int __ieee802_11_set_beacon(struct hostapd_data *hapd)
 #endif /* CONFIG_IEEE80211BE */
 
 	if (cmode &&
+#ifndef CONFIG_VENDOR_MXL
 	    hostapd_set_freq_params(&freq, iconf->hw_mode, iface->freq,
 				    iconf->channel, iconf->enable_edmg,
 				    iconf->edmg_channel, iconf->ieee80211n,
@@ -3260,6 +3257,20 @@ static int __ieee802_11_set_beacon(struct hostapd_data *hapd)
 				    &cmode->he_capab[IEEE80211_MODE_AP],
 				    &cmode->eht_capab[IEEE80211_MODE_AP],
 				    hostapd_get_punct_bitmap(hapd)) == 0) {
+#else /* CONFIG_VENDOR_MXL */
+	    hostapd_set_freq_params(&freq, iconf->hw_mode, iface->freq,
+				    iconf->channel, iconf->enable_edmg,
+				    iconf->edmg_channel, iconf->ieee80211n,
+				    iconf->ieee80211ac, iconf->ieee80211ax,
+				    iconf->ieee80211be,
+				    iconf->secondary_channel,
+				    hostapd_get_oper_chwidth(iconf),
+				    hostapd_get_oper_centr_freq_seg0_idx(iconf),
+				    hostapd_get_oper_centr_freq_seg1_idx(iconf),
+				    cmode->vht_capab,
+				    &cmode->he_capab[IEEE80211_MODE_AP],
+				    &cmode->eht_capab[IEEE80211_MODE_AP]) == 0) {
+#endif /* CONFIG_VENDOR_MXL */
 		freq.link_id = -1;
 #ifdef CONFIG_IEEE80211BE
 		if (hapd->conf->mld_ap)
diff --git a/src/ap/dfs.c b/src/ap/dfs.c
index d72d58fc5..07b67512f 100644
--- a/src/ap/dfs.c
+++ b/src/ap/dfs.c
@@ -1125,6 +1125,7 @@ static int hostapd_dfs_request_channel_switch(struct hostapd_iface *iface,
 	if (iface->mconf)
 		ieee80211_mode = IEEE80211_MODE_MESH;
 #endif /* CONFIG_MESH */
+#ifndef CONFIG_VENDOR_MXL
 	err = hostapd_set_freq_params(&csa_settings.freq_params,
 				      iface->conf->hw_mode,
 				      freq, channel,
@@ -1142,6 +1143,24 @@ static int hostapd_dfs_request_channel_switch(struct hostapd_iface *iface,
 				      &cmode->he_capab[ieee80211_mode],
 				      &cmode->eht_capab[ieee80211_mode],
 				      hostapd_get_punct_bitmap(iface->bss[0]));
+#else /* CONFIG_VENDOR_MXL */
+	err = hostapd_set_freq_params(&csa_settings.freq_params,
+				      iface->conf->hw_mode,
+				      freq, channel,
+				      iface->conf->enable_edmg,
+				      iface->conf->edmg_channel,
+				      iface->conf->ieee80211n,
+				      iface->conf->ieee80211ac,
+				      iface->conf->ieee80211ax,
+				      iface->conf->ieee80211be,
+				      secondary_channel,
+				      new_vht_oper_chwidth,
+				      oper_centr_freq_seg0_idx,
+				      oper_centr_freq_seg1_idx,
+				      cmode->vht_capab,
+				      &cmode->he_capab[ieee80211_mode],
+				      &cmode->eht_capab[ieee80211_mode]);
+#endif /* CONFIG_VENDOR_MXL */
 
 	if (err) {
 		wpa_printf(MSG_ERROR,
diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index 3ea54a432..388bf041a 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -2108,7 +2108,7 @@ static int hostapd_mgmt_rx(struct hostapd_data *hapd, struct rx_mgmt *rx_mgmt)
 			struct ieee80211_mgmt *mgmt = (struct ieee80211_mgmt *) rx_mgmt->frame;
 
 			bssid = mgmt->da;
-			hapd = get_hapd_bssid(iface, bssid);
+			hapd = get_hapd_bssid(iface, bssid, rx_mgmt->link_id);
 
 			if ((hapd == NULL) || (hapd == HAPD_BROADCAST)) {
 				return 0;
@@ -3267,11 +3267,13 @@ void wpa_supplicant_event(void *ctx, enum wpa_event_type event,
 		break;
 #endif /* CONFIG_IEEE80211AX */
 #ifdef CONFIG_IEEE80211BE
+#ifndef CONFIG_VENDOR_MXL
 	case EVENT_MLD_INTERFACE_FREED:
 		wpa_printf(MSG_DEBUG, "MLD: Interface %s freed",
 			   hapd->conf->iface);
 		hostapd_mld_interface_freed(hapd);
 		break;
+#endif /* CONFIG_VENDOR_MXL */
 #endif /* CONFIG_IEEE80211BE */
 #ifdef CONFIG_VENDOR_MXL
 	case EVENT_LTQ_UNCONNECTED_STA_RX:
@@ -3319,7 +3321,7 @@ void wpa_supplicant_event_global(void *ctx, enum wpa_event_type event,
 #ifdef CONFIG_VENDOR_MXL
 /* wrapper to access static functions from MXL code */
 struct hostapd_channel_data * mxl_hostapd_get_mode_channel(struct hostapd_iface *iface, unsigned int freq) { return hostapd_get_mode_channel(iface, freq); }
-struct hostapd_data * mxl_get_hapd_bssid(struct hostapd_iface *iface, const u8 *bssid) { return get_hapd_bssid(iface, bssid); }
+struct hostapd_data * mxl_get_hapd_bssid(struct hostapd_iface *iface, const u8 *bssid) { return get_hapd_bssid(iface, bssid, -1); }
 _Static_assert(HAPD_BROADCAST == MXL_HAPD_BROADCAST, "open source and MXL macros not aligned!");
 
 u16 hostapd_get_beacon_int(const struct hostapd_data *hapd)
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 419b59feb..42f9115a3 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -728,6 +728,7 @@ void hostapd_free_hapd_data(struct hostapd_data *hapd)
 			 * driver wrapper may have removed its internal instance
 			 * and hapd->drv_priv is not valid anymore.
 			 */
+
 			hapd->drv_priv = NULL;
 		}
 	}
@@ -6159,6 +6160,7 @@ static int hostapd_change_config_freq(struct hostapd_data *hapd,
 
 	/* if a pointer to old_params is provided we save previous state */
 	if (old_params &&
+#ifndef CONFIG_VENDOR_MXL
 	    hostapd_set_freq_params(old_params, conf->hw_mode,
 				    hostapd_hw_get_freq(hapd, conf->channel),
 				    conf->channel, conf->enable_edmg,
@@ -6174,6 +6176,22 @@ static int hostapd_change_config_freq(struct hostapd_data *hapd,
 				    mode ? &mode->eht_capab[IEEE80211_MODE_AP] :
 				    NULL,
 				    hostapd_get_punct_bitmap(hapd)))
+#else /* CONFIG_VENDOR_MXL */
+	    hostapd_set_freq_params(old_params, conf->hw_mode,
+				    hostapd_hw_get_freq(hapd, conf->channel),
+				    conf->channel, conf->enable_edmg,
+				    conf->edmg_channel, conf->ieee80211n,
+				    conf->ieee80211ac, conf->ieee80211ax,
+				    conf->ieee80211be, conf->secondary_channel,
+				    hostapd_get_oper_chwidth(conf),
+				    hostapd_get_oper_centr_freq_seg0_idx(conf),
+				    hostapd_get_oper_centr_freq_seg1_idx(conf),
+				    conf->vht_capab,
+				    mode ? &mode->he_capab[IEEE80211_MODE_AP] :
+				    NULL,
+				    mode ? &mode->eht_capab[IEEE80211_MODE_AP] :
+				    NULL))
+#endif /* CONFIG_VENDOR_MXL */
 		return -1;
 
 	switch (params->bandwidth) {
@@ -7321,6 +7339,9 @@ int mxl_hostapd_remove_bss(struct hostapd_iface *iface, unsigned int idx)
 
 void mxl_hostapd_clear_drv_priv(struct hostapd_data *hapd)
 {
-	return hostapd_clear_drv_priv(hapd);
+	struct hostapd_iface *iface = hapd->iface;
+	const struct wpa_driver_ops *driver = iface->bss[0]->driver;
+	void *drv_priv = iface->bss[0]->drv_priv;
+	return hostapd_cleanup_driver(driver, drv_priv, iface);
 }
 #endif /* CONFIG_VENDOR_MXL */
diff --git a/src/ap/hostapd.h b/src/ap/hostapd.h
index 584343d98..1313c0ad9 100644
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -24,7 +24,6 @@
 #ifdef UBUS_SUPPORT
 #include "ubus.h"
 #endif /* UBUS_SUPPORT */
-#endif /* CONFIG_VENDOR_MXL */
 
 #define OCE_STA_CFON_ENABLED(hapd) \
 	((hapd->conf->oce & OCE_STA_CFON) && \
@@ -33,7 +32,6 @@
 	((hapd->conf->oce & OCE_AP) && \
 	 (hapd->iface->drv_flags & WPA_DRIVER_FLAGS_OCE_AP))
 
-#ifdef CONFIG_VENDOR_MXL
 typedef enum {
 	GUARD_INTERVAL_400	= 0x01,
 	GUARD_INTERVAL_800	= 0x02,
@@ -55,6 +53,7 @@ struct wpa_ctrl_dst;
 struct radius_server_data;
 struct upnp_wps_device_sm;
 struct hostapd_data;
+struct mld_link_info;
 struct sta_info;
 struct ieee80211_ht_capabilities;
 struct full_dynamic_vlan;
@@ -211,21 +210,6 @@ struct hostapd_sae_commit_queue {
 	u8 msg[];
 };
 
-struct mld_link_info {
-	u8 valid:1;
-	u8 nstr_bitmap_len:2;
-	u8 local_addr[ETH_ALEN];
-	u8 peer_addr[ETH_ALEN];
-
-	u8 nstr_bitmap[2];
-
-	u16 capability;
-
-	u16 status;
-	u16 resp_sta_profile_len;
-	u8 *resp_sta_profile;
-};
-
 /**
  * struct hostapd_data - hostapd per-BSS data structure
  */
@@ -1075,13 +1059,6 @@ static inline bool hostapd_mld_is_first_bss(struct hostapd_data *hapd)
 
 u16 hostapd_get_punct_bitmap(struct hostapd_data *hapd);
 
-static inline bool ap_pmf_enabled(struct hostapd_bss_config *conf)
-{
-	return conf->ieee80211w != NO_MGMT_FRAME_PROTECTION ||
-		conf->rsn_override_mfp != NO_MGMT_FRAME_PROTECTION ||
-		conf->rsn_override_mfp_2 != NO_MGMT_FRAME_PROTECTION;
-}
-
 #ifdef CONFIG_VENDOR_MXL
 void hostapd_tx_queue_params(struct hostapd_data *hapd);
 void hostapd_event_ltq_report_event(struct hostapd_data *hapd,
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 197a88b6c..aea74aa65 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -4302,7 +4302,6 @@ static u16 owe_process_assoc_req(struct hostapd_data *hapd,
 
 	wpa_hexdump_key(MSG_DEBUG, "OWE: PMK", sta->owe_pmk, sta->owe_pmk_len);
 	wpa_hexdump(MSG_DEBUG, "OWE: PMKID", pmkid, PMKID_LEN);
-#ifdef CONFIG_VENDOR_MXL
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 	u8 *spa_addr = sta->addr;
 
@@ -4311,23 +4310,14 @@ static u16 owe_process_assoc_req(struct hostapd_data *hapd,
 		spa_addr = sta->mxl_sta_info.non_ap_mld_mac;
 		if (sta->mxl_sta_info.linked_sta) {
 			wpa_auth_pmksa_add2(hapd->mxl_data.sibling_hapd->wpa_auth, spa_addr, sta->owe_pmk,
-					    sta->owe_pmk_len, pmkid, 0, WPA_KEY_MGMT_OWE, NULL);
+					    sta->owe_pmk_len, pmkid, 0, WPA_KEY_MGMT_OWE, NULL, false);
 		}
 	}
-
-	wpa_auth_pmksa_add2(hapd->wpa_auth, spa->addr, sta->owe_pmk,
-			    sta->owe_pmk_len, pmkid, 0, WPA_KEY_MGMT_OWE,
-			    NULL, ap_sta_is_mld(hapd, sta));
 #else /* CONFIG_IEEE80211BE_MXL_MLO */
 	wpa_auth_pmksa_add2(hapd->wpa_auth, sta->addr, sta->owe_pmk,
 			    sta->owe_pmk_len, pmkid, 0, WPA_KEY_MGMT_OWE,
 			    NULL, ap_sta_is_mld(hapd, sta));
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
-#else /* CONFIG_VENDOR_MXL */
-	wpa_auth_pmksa_add2(hapd->wpa_auth, sta->addr, sta->owe_pmk,
-			    sta->owe_pmk_len, pmkid, 0, WPA_KEY_MGMT_OWE,
-			    NULL, ap_sta_is_mld(hapd, sta));
-#endif /* CONFIG_VENDOR_MXL */
 
 	return WLAN_STATUS_SUCCESS;
 }
@@ -4465,6 +4455,7 @@ end:
 #endif /* CONFIG_OWE */
 
 
+#ifndef CONFIG_VENDOR_MXL
 static bool hapd_is_known_sta(struct hostapd_data *hapd, struct sta_info *sta,
 			      const u8 *ies, size_t ies_len)
 {
@@ -4513,6 +4504,7 @@ static bool hapd_is_known_sta(struct hostapd_data *hapd, struct sta_info *sta,
 
 	return true;
 }
+#endif /* CONFIG_VENDOR_MXL */
 
 
 static bool check_sa_query(struct hostapd_data *hapd, struct sta_info *sta,
@@ -5508,7 +5500,9 @@ static int add_associated_sta(struct hostapd_data *hapd,
 	int set = 1;
 	const u8 *mld_link_addr = NULL;
 	bool mld_link_sta = false;
+#ifndef CONFIG_VENDOR_MXL
 	u16 eml_cap = 0;
+#endif /* CONFIG_VENDOR_MXL */
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 	if (mxl_is_mlo_enabled(hapd->conf) && mxl_mlo_is_ml_sta(sta)) {
@@ -5525,7 +5519,9 @@ static int add_associated_sta(struct hostapd_data *hapd,
 
 		if (hapd->mld_link_id != sta->mld_assoc_link_id)
 			set = 0;
+#ifndef CONFIG_VENDOR_MXL
 		eml_cap = sta->mld_info.common_info.eml_capa;
+#endif /* CONFIG_VENDOR_MXL */
 	}
 #endif /* CONFIG_IEEE80211BE */
 
@@ -9052,18 +9048,7 @@ static size_t hostapd_eid_rnr_colocation_len(struct hostapd_data *hapd,
 		if (!iface || iface == hapd->iface ||
 		    iface->state != HAPD_IFACE_ENABLED ||
 		    !is_6ghz_op_class(iface->conf->op_class))
-#ifdef CONFIG_VENDOR_MXL
-#ifdef CONFIG_IEEE80211BE_MXL_MLO
-		if (!iface)
-			continue;
-#endif /* CONFIG_IEEE80211BE_MXL_MLO */
-#endif /* CONFIG_VENDOR_MXL */
-
-#ifdef CONFIG_IEEE80211BE
-		if (hostapd_is_ml_partner(hapd, iface->bss[0]))
-			ap_mld = true;
-#endif /* CONFIG_IEEE80211BE */
-			continue;
+				continue;
 
 		len += hostapd_eid_rnr_iface_len(iface->bss[0], hapd,
 						 current_len, NULL, false);
@@ -9362,7 +9347,7 @@ static u8 * hostapd_eid_rnr_iface(struct hostapd_data *hapd,
 
 #ifdef CONFIG_VENDOR_MXL
 	/* Skip writing to eid buffer even rnr header if length is zero */
-	if (!hostapd_eid_rnr_iface_len(hapd, reporting_hapd, &rnr_len, NULL))
+	if (!hostapd_eid_rnr_iface_len(hapd, reporting_hapd, &rnr_len, NULL, false))
 		return eid;
 #endif /* CONFIG_VENDOR_MXL */
 
@@ -9437,18 +9422,7 @@ static u8 * hostapd_eid_rnr_colocation(struct hostapd_data *hapd, u8 *eid,
 		if (!iface || iface == hapd->iface ||
 		    iface->state != HAPD_IFACE_ENABLED ||
 		    !is_6ghz_op_class(iface->conf->op_class))
-#ifdef CONFIG_VENDOR_MXL
-#ifdef CONFIG_IEEE80211BE_MXL_MLO
-		if (!iface)
-			continue;
-#endif /* CONFIG_IEEE80211BE_MXL_MLO */
-#endif /* CONFIG_VENDOR_MXL */
-
-#ifdef CONFIG_IEEE80211BE
-		if (hostapd_is_ml_partner(hapd, iface->bss[0]))
-			ap_mld = true;
-#endif /* CONFIG_IEEE80211BE */
-			continue;
+				continue;
 
 		eid = hostapd_eid_rnr_iface(iface->bss[0], hapd, eid,
 					    current_len, NULL, false);
diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index 53ac62869..ca0899645 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -1427,7 +1427,7 @@ u16 hostapd_process_ml_assoc_req(struct hostapd_data *hapd,
 
 #ifdef CONFIG_VENDOR_MXL
       wpa_printf(MSG_DEBUG, "MLD: link ctrl=0x%x, " MACSTR
-            ", nstr bitmap len=%zu",
+            ", nstr bitmap len=%u",
             control, MAC2STR(link_info->peer_addr),
             link_info->nstr_bitmap_len);
 #else
diff --git a/src/ap/ieee802_11_ht.c b/src/ap/ieee802_11_ht.c
index c2ca9f86f..20d704361 100644
--- a/src/ap/ieee802_11_ht.c
+++ b/src/ap/ieee802_11_ht.c
@@ -117,55 +117,6 @@ u8 * hostapd_eid_ht_capabilities(struct hostapd_data *hapd, u8 *eid)
 	return pos;
 }
 
-#ifdef CONFIG_VENDOR_MXL
-static void set_ht_param(struct hostapd_data *hapd, struct ieee80211_ht_operation *oper)
-{
-	int secondary_channel = hapd->iconf->secondary_channel;
-#ifdef CONFIG_IEEE80211BE
-		enum oper_chan_width chwidth = hostapd_get_oper_chwidth(hapd->iconf);
-		u16 bw = 0, punct_bitmap = hapd->iconf->punct_bitmap;
-		u8 offset, chan_bit_pos;
-	
-		switch (chwidth) {
-		case CONF_OPER_CHWIDTH_80MHZ:
-			bw = 80;
-			offset = 6;
-			break;
-		case CONF_OPER_CHWIDTH_160MHZ:
-			bw = 160;
-			offset = 14;
-			break;
-		case CONF_OPER_CHWIDTH_320MHZ:
-			bw = 320;
-			offset = 30;
-			break;
-		default:
-			goto no_update;
-		}
-	
-		chan_bit_pos = (hapd->iconf->channel -
-				hostapd_get_oper_centr_freq_seg0_idx(hapd->iconf) +
-				offset) / 4;
-		/* Check if secondary channel is punctured */
-		if (bw >= 80 && punct_bitmap && secondary_channel &&
-		    (punct_bitmap & BIT(chan_bit_pos + secondary_channel)))
-			return; /* Do not indicate punctured secondary channel for HT */
-no_update:
-#endif /* CONFIG_IEEE80211BE */
-	
-		if (secondary_channel == 1)
-			oper->ht_param |= HT_INFO_HT_PARAM_SECONDARY_CHNL_ABOVE |
-				HT_INFO_HT_PARAM_STA_CHNL_WIDTH;
-		if (secondary_channel == -1)
-			oper->ht_param |= HT_INFO_HT_PARAM_SECONDARY_CHNL_BELOW |
-				HT_INFO_HT_PARAM_STA_CHNL_WIDTH;
-#ifdef CONFIG_VENDOR_MXL
-		if (hapd->iconf->ht_rifs)
-			oper->ht_param |= HT_INFO_HT_PARAM_RIFS_MODE;
-#endif /* CONFIG_VENDOR_MXL */
-}
-#endif /* CONFIG_VENDOR_MXL */
-
 
 static void set_ht_param(struct hostapd_data *hapd,
 			 struct ieee80211_ht_operation *oper)
diff --git a/src/ap/ieee802_11_shared.c b/src/ap/ieee802_11_shared.c
index b52c21334..e77bcb5b4 100644
--- a/src/ap/ieee802_11_shared.c
+++ b/src/ap/ieee802_11_shared.c
@@ -397,9 +397,6 @@ static void hostapd_ext_capab_byte(struct hostapd_data *hapd, u8 *pos, int idx,
 #ifdef CONFIG_VENDOR_MXL
 	u8 ext_cap = 0;
 
-#ifdef CONFIG_SCS_MXL
-	struct ieee80211_eht_capabilities *eht_cap = &hapd->iconf->mxl_conf.eht_capab_dbg;
-#endif /* CONFIG_SCS_MXL */
 	if (idx < hapd->iface->extended_capa_len)
 		ext_cap = hapd->iface->extended_capa[idx];
 #endif /* CONFIG_VENDOR_MXL */
@@ -1728,7 +1725,7 @@ u8 * hostapd_eid_multiple_bssid(struct hostapd_data *hapd, u8 *eid, bool is_beac
 		 */
 		if (bss->conf->wpa_ie_required_for_mbss) {
 			/* create wpa_auth_conf from bss configuration */
-			hostapd_wpa_auth_conf(bss->conf, bss->iconf, wpa_auth_conf);
+			hostapd_wpa_auth_conf(hapd->iface, bss->conf, bss->iconf, wpa_auth_conf);
 			if (wpa_auth_conf_gen_multibss_wpa_ie(wpa_auth_conf, &wpa_ie, &wpa_ie_len) < 0) {
 				wpa_printf(MSG_ERROR,"Multibss WPA-IE create failed !!!");
 				eid = NULL;
diff --git a/src/ap/mxl_beacon.c b/src/ap/mxl_beacon.c
index bdf4c824f..af7093af6 100644
--- a/src/ap/mxl_beacon.c
+++ b/src/ap/mxl_beacon.c
@@ -16,7 +16,6 @@
 #include "common/ieee802_11_defs.h"
 #include "common/ieee802_11_common.h"
 #include "ieee802_11.h"
-#include "mxl_mld.h"
 #include "hostapd.h"
 #include "ap_config.h"
 #include "ap_drv_ops.h"
@@ -26,6 +25,7 @@
 #include "wmm.h"
 #include "wpa_auth.h"
 #include "wpa_auth_glue.h"
+#include "mxl_mld.h"
 #include "mxl_beacon.h"
 #include "crc32.h"
 
diff --git a/src/ap/mxl_drv_callbacks.c b/src/ap/mxl_drv_callbacks.c
index 78853c608..a06ef19ff 100644
--- a/src/ap/mxl_drv_callbacks.c
+++ b/src/ap/mxl_drv_callbacks.c
@@ -24,6 +24,7 @@
 #include "ap_drv_ops.h"
 #include "beacon.h"
 #include "drivers/mxl_driver.h"
+#include "mxl_dfs_common.h"
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 void mxl_hostapd_notif_disassoc(struct hostapd_data *hapd, const u8 *addr)
diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index 6b07b41b6..6a9756ce1 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -2905,8 +2905,6 @@ void mxl_mlo_hostapd_interface_deinit(struct hostapd_iface *iface)
 	wpa_printf(MSG_DEBUG, "%s: driver=%p drv_priv=%p -> hapd_deinit",
 			__func__, driver, drv_priv);
 	if (driver && driver->hapd_deinit && drv_priv) {
-		if (!iface->bss[0]->mld_first_bss)
-			driver->hapd_deinit(drv_priv);
 		mxl_hostapd_clear_drv_priv(iface->bss[0]);
 	}
 	hostapd_interface_free(iface);
@@ -2980,7 +2978,7 @@ int mxl_switch_channel(struct hostapd_data *hapd, struct csa_settings *settings)
 	}
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
 	if (hapd->iconf->ieee80211be) {
-		mxl_hostapd_set_eht_punct_bitmap(hapd, settings->punct_bitmap);
+		mxl_hostapd_set_eht_punct_bitmap(hapd, hapd->iconf->punct_bitmap);
 
 		for (i = 1; i < iface->num_bss; i++) {
 			hapd_csa = iface->bss[i];
@@ -3000,7 +2998,7 @@ int mxl_switch_channel(struct hostapd_data *hapd, struct csa_settings *settings)
 				}
 			}
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
-			mxl_hostapd_set_eht_punct_bitmap(hapd_csa, settings->punct_bitmap);
+			mxl_hostapd_set_eht_punct_bitmap(hapd_csa, hapd->iconf->punct_bitmap);
 		}
 	}
 #endif /* defined(CONFIG_IEEE80211BE) */
diff --git a/src/ap/mxl_ieee802_11.c b/src/ap/mxl_ieee802_11.c
index eeb0ef5f9..8f46a040f 100644
--- a/src/ap/mxl_ieee802_11.c
+++ b/src/ap/mxl_ieee802_11.c
@@ -428,7 +428,7 @@ u16 mxl_mlo_check_assoc_ie(struct hostapd_data *hapd, struct sta_info *sta, cons
 			goto out;
 		}
 
-		if (ieee802_11_parse_link_assoc_req(pos, left, &elems, mlbuf,
+		if (ieee802_11_parse_link_assoc_req(&elems, mlbuf,
 		    hapd->mxl_data.sibling_hapd->conf->mxl_bss_conf.link_id, true) ==
 				ParseFailed) {
 			wpa_printf(MSG_DEBUG,
diff --git a/src/ap/mxl_mld.c b/src/ap/mxl_mld.c
index 87c690669..da9c807eb 100644
--- a/src/ap/mxl_mld.c
+++ b/src/ap/mxl_mld.c
@@ -18,7 +18,6 @@
 #include "common/mxl_ieee802_11_common.h"
 #include "common/mxl_sae.h"
 #include "ieee802_11.h"
-#include "mxl_mld.h"
 #include "hostapd.h"
 #include "ap_config.h"
 #include "ap_drv_ops.h"
@@ -30,10 +29,13 @@
 #include "wpa_auth_glue.h"
 #include "mxl_wpa_auth_glue.h"
 #include "mxl_wds.h"
+#include "mxl_mld.h"
 #include "mxl_beacon.h"
 #include "mxl_wmm.h"
 #include "mxl_ieee802_11.h"
 
+#define IS_EMPTY_MAC(mac) (memcmp((mac), "\x00\x00\x00\x00\x00\x00", 6) == 0)
+
 static void mxl_set_link_id(struct hostapd_data *hapd)
 {
 	if (is_6ghz_op_class(hapd->iconf->op_class))
@@ -762,7 +764,7 @@ static u8 * mxl_hostapd_eid_mld_staprof_probe(struct hostapd_data *hapd, u8 *eid
 #endif /* CONFIG_IEEE80211AX */
 	if ((hapd->iconf->ieee80211ac && !hapd->conf->disable_11ac) ||
 			(hapd->iconf->ieee80211ax && !hapd->conf->disable_11ax))
-		eid = hostapd_eid_wb_chsw_wrapper(hapd, eid);
+		eid = hostapd_eid_chsw_wrapper(hapd, eid);
 
 #ifdef CONFIG_IEEE80211AX
 	if (hapd->iconf->ieee80211ax && !hapd->conf->disable_11ax) {
@@ -1524,7 +1526,7 @@ int mxl_ml_sta_add(struct hostapd_data *hapd, struct sta_info *sta)
 
 	if (((!(sta->mxl_sta_info.ext_flags & WLAN_STA_SINGLE_LINK_ML_ASSOC)) &&
 				sta->mxl_sta_info.linked_sta == NULL) ||
-			sta->mxl_sta_info.non_ap_mld_mac == NULL) {
+			IS_EMPTY_MAC(sta->mxl_sta_info.non_ap_mld_mac)) {
 		os_free(ml_sta_params);
 		return -1;
 	}
@@ -1784,7 +1786,7 @@ int mxl_ml_handle_csa_critical_update(struct hostapd_data *hapd, struct csa_sett
 		sibling_hapd->mxl_data.ml_ie_csw_ie_len_add = true;
 	}
 
-	if ((settings->punct_bitmap) || (settings->freq_params.bandwidth == 320)) {
+	if ((settings->freq_params.punct_bitmap) || (settings->freq_params.bandwidth == 320)) {
 		sibling_hapd->mxl_data.bw_indication_ie_len_add = true;
 	}
 
diff --git a/src/ap/mxl_mld.h b/src/ap/mxl_mld.h
index 6143104c1..5c5db1d9e 100644
--- a/src/ap/mxl_mld.h
+++ b/src/ap/mxl_mld.h
@@ -132,6 +132,8 @@ enum multilink_type {
 #endif /* CONFIG_EPCS_MXL */
 };
 
+struct sta_info;
+
 static inline bool mxl_is_mlo_enabled(struct hostapd_bss_config *conf)
 {
 	return (conf->mxl_bss_conf.mlo_enable != 0);
diff --git a/src/ap/mxl_utils.c b/src/ap/mxl_utils.c
index be95eabdc..4bdfedf62 100644
--- a/src/ap/mxl_utils.c
+++ b/src/ap/mxl_utils.c
@@ -16,8 +16,8 @@
 #include "common.h"
 #include "common/ieee802_11_defs.h"
 #include "fst/fst.h"
-#include "sta_info.h"
 #include "hostapd.h"
+#include "sta_info.h"
 #include "common/ptksa_cache.h"
 #include "wpa_auth.h"
 #include "ap_mlme.h"
diff --git a/src/ap/mxl_wpa_auth.c b/src/ap/mxl_wpa_auth.c
index 29f1f5960..4a7bb6803 100644
--- a/src/ap/mxl_wpa_auth.c
+++ b/src/ap/mxl_wpa_auth.c
@@ -34,13 +34,13 @@
 #include "drivers/driver.h"
 #include "ap_config.h"
 #include "ieee802_11.h"
+#include "hostapd.h"
 #include "sta_info.h"
 #include "wpa_auth.h"
 #include "mxl_wpa_auth.h"
 #include "pmksa_cache_auth.h"
 #include "wpa_auth_i.h"
 #include "wpa_auth_ie.h"
-#include "hostapd.h"
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 #define PN_LENGTH 6
@@ -313,7 +313,7 @@ void mxl_ml_wpa_ptkinitnegotiating(struct wpa_state_machine *sm)
 		secure = 1;
 		gtk_len = gsm->GTK_len;
 		if (sm->wpa_auth->conf.disable_gtk ||
-		    sm->wpa_key_mgmt == WPA_KEY_MGMT_OSEN) {
+		    sm->wpa_key_mgmt == WPA_KEY_MGMT_OWE) {
 			/*
 			 * Provide unique random GTK to each STA to prevent use
 			 * of GTK in the BSS.
@@ -390,7 +390,7 @@ void mxl_ml_wpa_ptkinitnegotiating(struct wpa_state_machine *sm)
 		size_t elen;
 
 		elen = pos - kde;
-		res = wpa_insert_pmkid(kde, &elen, sm->pmk_r1_name);
+		res = wpa_insert_pmkid(kde, &elen, sm->pmk_r1_name, true);
 		if (res < 0) {
 			wpa_printf(MSG_ERROR, "FT: Failed to insert "
 				   "PMKR1Name into RSN IE in EAPOL-Key data");
@@ -553,7 +553,7 @@ void mxl_ml_wpa_ptk_group_rekeynegotiating(struct wpa_state_machine *sm)
 
 	gtk = gsm->GTK[gsm->GN - 1];
 	if (sm->wpa_auth->conf.disable_gtk ||
-	    sm->wpa_key_mgmt == WPA_KEY_MGMT_OSEN) {
+	    sm->wpa_key_mgmt == WPA_KEY_MGMT_OWE) {
 		/*
 		 * Provide unique random GTK to each STA to prevent use
 		 * of GTK in the BSS.
@@ -678,8 +678,7 @@ void mxl_ml_wpa_ptkstart (struct wpa_state_machine *sm)
 	if (sm->wpa == WPA_VERSION_WPA2 &&
 			(wpa_key_mgmt_wpa_ieee8021x(sm->wpa_key_mgmt) ||
 			 (sm->wpa_key_mgmt == WPA_KEY_MGMT_OWE && sm->pmksa) ||
-			 wpa_key_mgmt_sae(sm->wpa_key_mgmt)) &&
-			sm->wpa_key_mgmt != WPA_KEY_MGMT_OSEN) {
+			 wpa_key_mgmt_sae(sm->wpa_key_mgmt))) {
 		pmkid = kde + kde_len;
 		pmkid_len = 2 + RSN_SELECTOR_LEN + PMKID_LEN;
 		pmkid[0] = WLAN_EID_VENDOR_SPECIFIC;
diff --git a/src/ap/pmksa_cache_auth.c b/src/ap/pmksa_cache_auth.c
index 9322b4e42..0276e364c 100644
--- a/src/ap/pmksa_cache_auth.c
+++ b/src/ap/pmksa_cache_auth.c
@@ -13,6 +13,7 @@
 #include "eapol_auth/eapol_auth_sm.h"
 #include "eapol_auth/eapol_auth_sm_i.h"
 #include "radius/radius_das.h"
+#include "hostapd.h"
 #include "sta_info.h"
 #include "ap_config.h"
 #include "pmksa_cache_auth.h"
diff --git a/src/ap/rrm.c b/src/ap/rrm.c
index 43a5b88f6..4602681dd 100644
--- a/src/ap/rrm.c
+++ b/src/ap/rrm.c
@@ -370,6 +370,7 @@ static void hostapd_link_mesr_rep_timeout_handler(void *eloop_data,
 }
 
 
+#ifndef CONFIG_VENDOR_MXL
 static void hostapd_handle_link_mesr_report(struct hostapd_data *hapd,
 					    const u8 *buf, size_t len)
 {
@@ -403,6 +404,7 @@ static void hostapd_handle_link_mesr_report(struct hostapd_data *hapd,
 	wpa_msg(hapd->msg_ctx, MSG_INFO, LINK_MSR_RESP_RX MACSTR " %u %s",
 		MAC2STR(mgmt->sa), report->dialog_token, report_msg);
 }
+#endif /* CONFIG_VENDOR_MXL */
 
 
 void hostapd_handle_radio_measurement(struct hostapd_data *hapd,
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index 9a020a816..5cdfa8b9a 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -204,6 +204,7 @@ void ap_free_sta_pasn(struct hostapd_data *hapd, struct sta_info *sta)
 #endif /* CONFIG_PASN */
 
 
+#ifndef CONFIG_VENDOR_MXL
 static void __ap_free_sta(struct hostapd_data *hapd, struct sta_info *sta)
 {
 #ifdef CONFIG_IEEE80211BE
@@ -214,6 +215,7 @@ static void __ap_free_sta(struct hostapd_data *hapd, struct sta_info *sta)
 
 	hostapd_drv_sta_remove(hapd, sta->addr);
 }
+#endif /* CONFIG_VENDOR_MXL */
 
 
 #ifdef CONFIG_IEEE80211BE
@@ -605,11 +607,11 @@ void ap_handle_timer(void *eloop_ctx, void *timeout_ctx)
 	unsigned long next_time = 0;
 #ifndef CONFIG_VENDOR_MXL
 	int reason;
-	int max_inactivity = hapd->conf->ap_max_inactivity;
 #else /* CONFIG_VENDOR_MXL */
 	u16 reason;
 	unsigned int disassoc_delay;
 #endif /* CONFIG_VENDOR_MXL */
+	int max_inactivity = hapd->conf->ap_max_inactivity;
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 	if (mxl_is_mlo_enabled(hapd->conf)) {
@@ -1131,7 +1133,6 @@ static void ap_sta_disconnect_common(struct hostapd_data *hapd,
 
 #ifdef CONFIG_VENDOR_MXL
 	sta->flags &= ~WLAN_STA_BSS_TRANS_MGMT_REQ_TIMEOUT;
-	sta->disassoc_reason = reason;
 	sta->disconnect_source = DISCONNECT_SOURCE_LOCAL;
 #endif /* CONFIG_VENDOR_MXL */
 	ap_sta_set_authorized(hapd, sta, 0);
@@ -1763,29 +1764,9 @@ const u8 * ap_sta_wpa_get_dpp_pkhash(struct hostapd_data *hapd,
 bool ap_sta_set_authorized_flag(struct hostapd_data *hapd, struct sta_info *sta,
 				int authorized)
 {
-#ifdef CONFIG_VENDOR_MXL
-	const u8 *dev_addr = NULL;
-	char *buf, *pos, *end;
-	int buflen = STA_EVENT_MAX_BUF_LEN;
-	char *sta_extra_cap = NULL;
-	int i, ret = 0, len = 0;
-	unsigned int max_ch_width = 0, phy_mode = 0;
-	unsigned int mumimo = 0, max_mcs = 0, max_streams = 0, tx_streams = 0, rx_streams = 0;
-	unsigned int rrm_cap = 0, band_2g_capable = 0, band_5g_capable = 0, band_6g_capable = 0;
-#ifdef CONFIG_P2P
-	u8 addr[ETH_ALEN];
-	u8 ip_addr_buf[4];
-#endif /* CONFIG_P2P */
-#endif /* CONFIG_VENDOR_MXL */
 	if (!!authorized == !!(sta->flags & WLAN_STA_AUTHORIZED))
 		return false;
 
-	buf = os_malloc(buflen);
-	if (!buf)
-		return;
-	pos = buf;
-	end = buf + buflen;
-
 	if (authorized) {
 		int mld_assoc_link_id = -1;
 
@@ -2188,6 +2169,7 @@ void ap_sta_set_authorized_event(struct hostapd_data *hapd,
 	char *buf, *pos, *end;
 	int buflen = STA_EVENT_MAX_BUF_LEN;
 	char *sta_extra_cap = NULL;
+	int mld_assoc_link_id = -1;
 	int i, ret = 0, len = 0;
 	unsigned int max_ch_width = 0, phy_mode = 0;
 	unsigned int mumimo = 0, max_mcs = 0, max_streams = 0, tx_streams = 0, rx_streams = 0;
@@ -2202,6 +2184,19 @@ void ap_sta_set_authorized_event(struct hostapd_data *hapd,
 	const u8 *ip_ptr = NULL;
 
 #ifdef CONFIG_VENDOR_MXL
+	buf = os_malloc(buflen);
+   if (!buf)
+           return;
+   pos = buf;
+   end = buf + buflen;
+
+	sta_extra_cap = os_malloc(STA_EXTRA_CAP_LEN);
+   if (sta_extra_cap == NULL) {
+      wpa_printf(MSG_ERROR, "malloc failed");
+      goto free_alloc;
+   }
+   sta_extra_cap[0] = '\0';
+
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 		if (mxl_mlo_is_ml_sta(sta)) {
 			/* Purge existing STA entry which matches with the current STA's MLD MAC */
@@ -3006,7 +3001,11 @@ int ap_sta_re_add(struct hostapd_data *hapd, struct sta_info *sta)
 {
 	const u8 *mld_link_addr = NULL;
 	bool mld_link_sta = false;
+#ifdef CONFIG_IEEE80211BE
+#ifndef CONFIG_VENDOR_MXL
 	u16 eml_cap = 0;
+#endif /* CONFIG_VENDOR_MXL */
+#endif /* CONFIG_IEEE80211BE */
 
 	/*
 	 * If a station that is already associated to the AP, is trying to
@@ -3022,7 +3021,9 @@ int ap_sta_re_add(struct hostapd_data *hapd, struct sta_info *sta)
 
 		mld_link_sta = sta->mld_assoc_link_id != mld_link_id;
 		mld_link_addr = sta->mld_info.links[mld_link_id].peer_addr;
+#ifndef CONFIG_VENDOR_MXL
 		eml_cap = sta->mld_info.common_info.eml_capa;
+#endif /* CONFIG_VENDOR_MXL */
 
 		/*
 		 * In case the AP is affiliated with an AP MLD, we need to
diff --git a/src/ap/sta_info.h b/src/ap/sta_info.h
index a663f14b3..07263d7ac 100644
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -67,6 +67,7 @@
 #endif /* CONFIG_VENDOR_MXL */
 
 struct hostapd_data;
+struct mld_link_info;
 
 struct mbo_non_pref_chan_info {
 	struct mbo_non_pref_chan_info *next;
@@ -494,7 +495,11 @@ static inline bool ap_sta_is_mld(struct hostapd_data *hapd,
 				 struct sta_info *sta)
 {
 #ifdef CONFIG_IEEE80211BE
+#ifndef CONFIG_VENDOR_MXL
 	return hapd->conf->mld_ap && sta && sta->mld_info.mld_sta;
+#else /* CONFIG_VENDOR_MXL */
+	return sta && sta->mld_info.mld_sta;
+#endif /* CONFIG_VENDOR_MXL */
 #else /* CONFIG_IEEE80211BE */
 	return false;
 #endif /* CONFIG_IEEE80211BE */
diff --git a/src/ap/utils.c b/src/ap/utils.c
index 1f359dcb1..6681b4c22 100644
--- a/src/ap/utils.c
+++ b/src/ap/utils.c
@@ -11,8 +11,8 @@
 #include "common.h"
 #include "common/ieee802_11_defs.h"
 #include "fst/fst.h"
-#include "sta_info.h"
 #include "hostapd.h"
+#include "sta_info.h"
 #ifdef CONFIG_VENDOR_MXL
 #include "ap/mxl_utils.h"
 #endif /* CONFIG_VENDOR_MXL */
diff --git a/src/ap/wds_wpa.c b/src/ap/wds_wpa.c
index 20964ee4b..a6d8d612a 100644
--- a/src/ap/wds_wpa.c
+++ b/src/ap/wds_wpa.c
@@ -431,7 +431,7 @@ int ltq_wds_wpa_add(struct hostapd_data *hapd, const u8 *src)
 		/* wpa_sm_set_param(sta->wpa, WPA_PARAM_PROTO, hapd->conf->wpa); */
 		wpa_sm_set_param(sta->wpa, WPA_PARAM_PROTO, WPA_PROTO_RSN);
 		wpa_sm_set_param(sta->wpa, WPA_PARAM_RSN_ENABLED,
-			!!(hapd->conf->wpa & (WPA_PROTO_RSN | WPA_PROTO_OSEN)));
+			!!(hapd->conf->wpa & WPA_PROTO_RSN));
 		wpa_sm_set_param(sta->wpa, WPA_PARAM_KEY_MGMT, hapd->conf->wpa_key_mgmt);
 		/* wpa_sm_set_param(sta->wpa, WPA_PARAM_PAIRWISE, hapd->conf->wpa_pairwise); */
 		wpa_sm_set_param(sta->wpa, WPA_PARAM_PAIRWISE, WPA_CIPHER_CCMP);
diff --git a/src/ap/wnm_ap.c b/src/ap/wnm_ap.c
index 1b8cf4cf3..2ab50bb12 100644
--- a/src/ap/wnm_ap.c
+++ b/src/ap/wnm_ap.c
@@ -1514,7 +1514,7 @@ int wnm_send_bss_tm_req(struct hostapd_data *hapd, struct sta_info *sta,
 			}
 		}
 #endif /* CONFIG_IEEE80211BE */
-
+	}
 #ifdef CONFIG_VENDOR_MXL
 	if (disassoc_timer && !(mxl_conf->ignore_da_timer)) {
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
diff --git a/src/ap/wpa_auth.c b/src/ap/wpa_auth.c
index a633e4aaf..2d95f4b5e 100644
--- a/src/ap/wpa_auth.c
+++ b/src/ap/wpa_auth.c
@@ -29,6 +29,7 @@
 #include "drivers/driver.h"
 #include "ap_config.h"
 #include "ieee802_11.h"
+#include "hostapd.h"
 #include "sta_info.h"
 #include "wpa_auth.h"
 #include "pmksa_cache_auth.h"
@@ -516,6 +517,7 @@ static void wpa_rekey_gmk(void *eloop_ctx, void *timeout_ctx)
 }
 
 
+#ifndef CONFIG_IEEE80211BE_MXL_MLO
 static void wpa_rekey_all_groups(struct wpa_authenticator *wpa_auth)
 {
 	struct wpa_group *group, *next;
@@ -584,6 +586,7 @@ static int wpa_rekey_all_groups_cb(struct wpa_authenticator *wpa_auth,
 }
 
 #endif /* CONFIG_IEEE80211BE */
+#endif /* CONFIG_IEEE80211BE_MXL_MLO */
 
 
 #ifndef CONFIG_IEEE80211BE_MXL_MLO
@@ -969,10 +972,12 @@ void wpa_deinit(struct wpa_authenticator *wpa_auth)
 			 * authenticator and start rekey timer.
 			 */
 			next_pa->primary_auth = true;
+#ifndef CONFIG_IEEE80211BE_MXL_MLO
 			if (next_pa->conf.wpa_group_rekey)
 				eloop_register_timeout(
 					next_pa->conf.wpa_group_rekey,
 					0, wpa_rekey_gtk, next_pa, NULL);
+#endif /* CONFIG_IEEE80211BE_MXL_MLO */
 		}
 	}
 #endif /* CONFIG_IEEE80211BE */
@@ -1170,14 +1175,18 @@ void wpa_auth_sta_deinit(struct wpa_state_machine *sm)
 
 	wpa_auth = sm->wpa_auth;
 	if (wpa_auth->conf.wpa_strict_rekey && sm->has_GTK) {
+#ifndef CONFIG_VENDOR_MXL
 		struct wpa_authenticator *primary_auth = wpa_auth;
+#endif /* CONFIG_VENDOR_MXL */
 
 		wpa_auth_logger(wpa_auth, wpa_auth_get_spa(sm), LOGGER_DEBUG,
 				"strict rekeying - force GTK rekey since STA is leaving");
 
 #ifdef CONFIG_IEEE80211BE
+#ifndef CONFIG_VENDOR_MXL
 		if (wpa_auth->is_ml && !wpa_auth->primary_auth)
 			primary_auth = wpa_get_primary_auth(wpa_auth);
+#endif /* CONFIG_VENDOR_MXL */
 #endif /* CONFIG_IEEE80211BE */
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
@@ -1656,12 +1665,6 @@ void wpa_receive(struct wpa_authenticator *wpa_auth,
 	if (!wpa_auth || !wpa_auth->conf.wpa || !sm)
 		return;
 
-#ifdef CONFIG_VENDOR_MXL
-#ifdef CONFIG_IEEE80211BE_MXL_MLO
-	is_mld = sm->mxl_sm.is_ml_sta;
-#endif /* CONFIG_IEEE80211BE_MXL_MLO */
-#endif /* CONFIG_VENDOR_MXL */
-
 	wpa_hexdump(MSG_MSGDUMP, "WPA: RX EAPOL data", data, data_len);
 
 	mic_len = wpa_mic_len(sm->wpa_key_mgmt, sm->pmk_len);
@@ -1951,11 +1954,11 @@ void wpa_receive(struct wpa_authenticator *wpa_auth,
 			wpa_auth_logger(wpa_auth, wpa_auth_get_spa(sm),
 					LOGGER_INFO,
 					"received EAPOL-Key Request for GTK rekeying");
-
+#ifndef CONFIG_IEEE80211BE_MXL_MLO
 			eloop_cancel_timeout(wpa_rekey_gtk,
 					     wpa_get_primary_auth(wpa_auth),
 					     NULL);
-#ifdef CONFIG_VENDOR_MXL
+#endif /* CONFIG_IEEE80211BE_MXL_MLO */
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 			eloop_cancel_timeout(mxl_wpa_rekey_gtk, wpa_auth, NULL);
 			if (wpa_auth_gtk_rekey_in_process(wpa_auth))
@@ -1971,16 +1974,8 @@ void wpa_receive(struct wpa_authenticator *wpa_auth,
 			else
 				wpa_rekey_gtk(wpa_get_primary_auth(wpa_auth),
 					      NULL);
-#endif
-#else /* CONFIG_VENDOR_MXL */
-			eloop_cancel_timeout(wpa_rekey_gtk, wpa_auth, NULL);
-			if (wpa_auth_gtk_rekey_in_process(wpa_auth))
-				wpa_auth_logger(wpa_auth, NULL, LOGGER_DEBUG,
-						"skip new GTK rekey - already in process");
-			else
-				wpa_rekey_gtk(wpa_get_primary_auth(wpa_auth),
-					      NULL);
-#endif /* CONFIG_VENDOR_MXL */
+#endif /* CONFIG_IEEE80211BE_MXL_MLO */
+		}
 	} else {
 		/* Do not allow the same key replay counter to be reused. */
 		wpa_replay_counter_mark_invalid(sm->key_replay,
@@ -6685,9 +6680,15 @@ int wpa_auth_pmksa_add_preauth(struct wpa_authenticator *wpa_auth,
 }
 
 
+#ifndef CONFIG_VENDOR_MXL
 int wpa_auth_pmksa_add_sae(struct wpa_authenticator *wpa_auth, const u8 *addr,
 			   const u8 *pmk, size_t pmk_len, const u8 *pmkid,
 			   int akmp, bool is_ml)
+#else /* CONFIG_VENDOR_MXL */
+int wpa_auth_pmksa_add_sae(struct wpa_authenticator *wpa_auth, const u8 *addr,
+			   const u8 *pmk, size_t pmk_len, const u8 *pmkid,
+			   int akmp)
+#endif /* CONFIG_VENDOR_MXL */
 {
 	struct rsn_pmksa_cache *pmksa = wpa_auth->pmksa;
 	const u8 *aa = wpa_auth->addr;
@@ -6699,12 +6700,14 @@ int wpa_auth_pmksa_add_sae(struct wpa_authenticator *wpa_auth, const u8 *addr,
 	if (!akmp)
 		akmp = WPA_KEY_MGMT_SAE;
 
+#ifndef CONFIG_VENDOR_MXL
 #ifdef CONFIG_IEEE80211BE
 	if (is_ml) {
 		pmksa = wpa_auth->ml_pmksa;
 		aa = wpa_auth->mld_addr;
 	}
 #endif /* CONFIG_IEEE80211BE */
+#endif /* CONFIG_VENDOR_MXL */
 
 	if (pmksa_cache_auth_add(pmksa, pmk, pmk_len, pmkid, NULL, 0, aa, addr,
 				 0, NULL, akmp))
diff --git a/src/ap/wpa_auth.h b/src/ap/wpa_auth.h
index 6665bb933..558e86789 100644
--- a/src/ap/wpa_auth.h
+++ b/src/ap/wpa_auth.h
@@ -534,9 +534,15 @@ int wpa_auth_pmksa_add_preauth(struct wpa_authenticator *wpa_auth,
 			       const u8 *pmk, size_t len, const u8 *sta_addr,
 			       int session_timeout,
 			       struct eapol_state_machine *eapol);
+#ifndef CONFIG_VENDOR_MXL
 int wpa_auth_pmksa_add_sae(struct wpa_authenticator *wpa_auth, const u8 *addr,
 			   const u8 *pmk, size_t pmk_len, const u8 *pmkid,
 			   int akmp, bool is_ml);
+#else /* CONFIG_VENDOR_MXL */
+int wpa_auth_pmksa_add_sae(struct wpa_authenticator *wpa_auth, const u8 *addr,
+			   const u8 *pmk, size_t pmk_len, const u8 *pmkid,
+			   int akmp);
+#endif /* CONFIG_VENDOR_MXL */
 void wpa_auth_add_sae_pmkid(struct wpa_state_machine *sm, const u8 *pmkid);
 int wpa_auth_pmksa_add2(struct wpa_authenticator *wpa_auth, const u8 *addr,
 			const u8 *pmk, size_t pmk_len, const u8 *pmkid,
diff --git a/src/ap/wpa_auth_glue.c b/src/ap/wpa_auth_glue.c
index f582c86f8..c6b9c0e8b 100644
--- a/src/ap/wpa_auth_glue.c
+++ b/src/ap/wpa_auth_glue.c
@@ -38,16 +38,14 @@
 #include "mxl_wpa_auth_glue.h"
 #endif /* CONFIG_VENDOR_MXL */
 
-#ifndef CONFIG_VENDOR_MXL
+#ifdef CONFIG_VENDOR_MXL
+void hostapd_wpa_auth_conf(struct hostapd_iface *iface,
+#else /* CONFIG_VENDOR_MXL */
 static void hostapd_wpa_auth_conf(struct hostapd_iface *iface,
+#endif /* CONFIG_VENDOR_MXL */
 				  struct hostapd_bss_config *conf,
 				  struct hostapd_config *iconf,
 				  struct wpa_auth_config *wconf)
-#else /* CONFIG_VENDOR_MXL */
-void hostapd_wpa_auth_conf(struct hostapd_bss_config *conf,
-				  struct hostapd_config *iconf,
-				  struct wpa_auth_config *wconf)
-#endif /* CONFIG_VENDOR_MXL */
 {
 	int sae_pw_id;
 
@@ -625,9 +623,9 @@ static int hostapd_wpa_auth_set_key(void *ctx, int vlan_id, enum wpa_alg alg,
 		hapd->last_gtk_len = key_len;
 	}
 #endif /* CONFIG_TESTING_OPTIONS */
+#endif /* CONFIG_VENDOR_MXL */
 	return hostapd_drv_set_key(ifname, hapd, alg, addr, idx, vlan_id,
 				   set_tx, NULL, 0, key, key_len, key_flag);
-#endif /* CONFIG_VENDOR_MXL */
 }
 
 
diff --git a/src/ap/wpa_auth_glue.h b/src/ap/wpa_auth_glue.h
index 5a21c063c..8d10e07f4 100644
--- a/src/ap/wpa_auth_glue.h
+++ b/src/ap/wpa_auth_glue.h
@@ -13,7 +13,8 @@ int hostapd_setup_wpa(struct hostapd_data *hapd);
 void hostapd_reconfig_wpa(struct hostapd_data *hapd);
 void hostapd_deinit_wpa(struct hostapd_data *hapd);
 #ifdef CONFIG_VENDOR_MXL
-void hostapd_wpa_auth_conf(struct hostapd_bss_config *conf,
+void hostapd_wpa_auth_conf(struct hostapd_iface *iface,
+				struct hostapd_bss_config *conf,
 				struct hostapd_config *iconf,
 				struct wpa_auth_config *wconf);
 #endif /* CONFIG_VENDOR_MXL */
diff --git a/src/ap/wpa_auth_ie.c b/src/ap/wpa_auth_ie.c
index 624d76d4f..a4e86d019 100644
--- a/src/ap/wpa_auth_ie.c
+++ b/src/ap/wpa_auth_ie.c
@@ -1617,15 +1617,11 @@ bool wpa_auth_write_fd_rsn_info(struct wpa_authenticator *wpa_auth,
 
 #endif /* CONFIG_FILS */
 
-#ifdef CONFIG_VENDOR_MXL
 int wpa_auth_conf_gen_multibss_wpa_ie(struct wpa_auth_config *conf, u8 **wpa_ie, size_t *wpa_ie_len)
 {
 	u8 *pos, buf[WPA_IE_MAX_LEN] = {0};
 	int res = -1;
 	pos = buf;
-	if (conf->wpa == WPA_PROTO_OSEN) {
-		pos = wpa_write_osen(conf, pos);
-	}
 	if (conf->wpa & WPA_PROTO_RSN) {
 		res = wpa_write_rsn_ie(conf,
 			pos, buf + sizeof(buf) - pos, NULL);
@@ -1688,4 +1684,3 @@ int wpa_modify_wpa_ie(struct wpa_state_machine *sm, u32 group_suite, u16 capab,
 	return 0;
 }
 #endif /* CONFIG_WDS_WPA */
-#endif /* CONFIG_VENDOR_MXL */
diff --git a/src/common/defs.h b/src/common/defs.h
index 0dfa15071..c8e229734 100644
--- a/src/common/defs.h
+++ b/src/common/defs.h
@@ -550,6 +550,20 @@ enum frame_encryption {
 };
 
 #define MAX_NUM_MLD_LINKS 15
+struct mld_link_info {
+   u8 valid:1;
+   u8 nstr_bitmap_len:2;
+   u8 local_addr[ETH_ALEN];
+   u8 peer_addr[ETH_ALEN];
+
+   u8 nstr_bitmap[2];
+
+   u16 capability;
+
+   u16 status;
+   u16 resp_sta_profile_len;
+   u8 *resp_sta_profile;
+};
 
 enum sae_pwe {
 	SAE_PWE_HUNT_AND_PECK = 0,
diff --git a/src/common/hw_features_common.c b/src/common/hw_features_common.c
index 41dccc23d..073af0708 100644
--- a/src/common/hw_features_common.c
+++ b/src/common/hw_features_common.c
@@ -494,6 +494,7 @@ void punct_update_legacy_bw(u16 bitmap, u8 pri, enum oper_chan_width *width,
 }
 
 
+#ifndef CONFIG_VENDOR_MXL
 int hostapd_set_freq_params(struct hostapd_freq_params *data,
 			    enum hostapd_hw_mode mode,
 			    int freq, int channel, int enable_edmg,
@@ -506,9 +507,24 @@ int hostapd_set_freq_params(struct hostapd_freq_params *data,
 			    struct he_capabilities *he_cap,
 			    struct eht_capabilities *eht_cap,
 			    u16 punct_bitmap)
+#else /* CONFIG_VENDOR_MXL */
+int hostapd_set_freq_params(struct hostapd_freq_params *data,
+			    enum hostapd_hw_mode mode,
+			    int freq, int channel, int enable_edmg,
+			    u8 edmg_channel, int ht_enabled,
+			    int vht_enabled, int he_enabled,
+			    bool eht_enabled, int sec_channel_offset,
+			    enum oper_chan_width oper_chwidth,
+			    int center_segment0,
+			    int center_segment1, u32 vht_caps,
+			    struct he_capabilities *he_cap,
+			    struct eht_capabilities *eht_cap)
+#endif /* CONFIG_VENDOR_MXL */
 {
 	enum oper_chan_width oper_chwidth_legacy;
+#ifndef CONFIG_VENDOR_MXL
 	u8 seg0_legacy, seg1_legacy;
+#endif /* CONFIG_VENDOR_MXL */
 
 	if (!he_cap || !he_cap->he_supported)
 		he_enabled = 0;
@@ -525,7 +541,9 @@ int hostapd_set_freq_params(struct hostapd_freq_params *data,
 	data->sec_channel_offset = sec_channel_offset;
 	data->center_freq1 = freq + sec_channel_offset * 10;
 	data->center_freq2 = 0;
+#ifndef CONFIG_VENDOR_MXL
 	data->punct_bitmap = punct_bitmap;
+#endif /* CONFIG_VENDOR_MXL */
 	if (oper_chwidth == CONF_OPER_CHWIDTH_80MHZ)
 		data->bandwidth = 80;
 	else if (oper_chwidth == CONF_OPER_CHWIDTH_160MHZ ||
@@ -698,12 +716,14 @@ int hostapd_set_freq_params(struct hostapd_freq_params *data,
 	}
 
 	oper_chwidth_legacy = oper_chwidth;
+#ifndef CONFIG_VENDOR_MXL
 	seg0_legacy = center_segment0;
 	seg1_legacy = center_segment1;
 	if (punct_bitmap)
 		punct_update_legacy_bw(punct_bitmap, channel,
 				       &oper_chwidth_legacy,
 				       &seg0_legacy, &seg1_legacy);
+#endif /* CONFIG_VENDOR_MXL */
 
 	if (data->eht_enabled || data->he_enabled ||
 	    data->vht_enabled) switch (oper_chwidth) {
diff --git a/src/common/hw_features_common.h b/src/common/hw_features_common.h
index 80e33adf2..2ef570ff8 100644
--- a/src/common/hw_features_common.h
+++ b/src/common/hw_features_common.h
@@ -37,6 +37,7 @@ int check_40mhz_2g4(struct hostapd_hw_modes *mode,
 		    int sec_chan);
 void punct_update_legacy_bw(u16 bitmap, u8 pri_chan,
 			    enum oper_chan_width *width, u8 *seg0, u8 *seg1);
+#ifndef CONFIG_VENDOR_MXL
 int hostapd_set_freq_params(struct hostapd_freq_params *data,
 			    enum hostapd_hw_mode mode,
 			    int freq, int channel, int edmg, u8 edmg_channel,
@@ -49,6 +50,19 @@ int hostapd_set_freq_params(struct hostapd_freq_params *data,
 			    struct he_capabilities *he_caps,
 			    struct eht_capabilities *eht_cap,
 			    u16 punct_bitmap);
+#else /* CONFIG_VENDOR_MXL */
+int hostapd_set_freq_params(struct hostapd_freq_params *data,
+             enum hostapd_hw_mode mode,
+             int freq, int channel, int edmg, u8 edmg_channel,
+             int ht_enabled,
+             int vht_enabled, int he_enabled,
+             bool eht_enabled, int sec_channel_offset,
+             enum oper_chan_width oper_chwidth,
+             int center_segment0,
+             int center_segment1, u32 vht_caps,
+             struct he_capabilities *he_caps,
+             struct eht_capabilities *eht_cap);
+#endif /* CONFIG_VENDOR_MXL */
 void set_disable_ht40(struct ieee80211_ht_capabilities *htcaps,
 		      int disabled);
 int ieee80211ac_cap_check(u32 hw, u32 conf);
diff --git a/src/common/ieee802_11_common.c b/src/common/ieee802_11_common.c
index b961a6dd1..68ccd3225 100644
--- a/src/common/ieee802_11_common.c
+++ b/src/common/ieee802_11_common.c
@@ -138,10 +138,12 @@ static int ieee802_11_parse_vendor_specific(const u8 *pos, size_t elen,
 			elems->sae_pk = pos + 4;
 			elems->sae_pk_len = elen - 4;
 			break;
+#ifndef CONFIG_VENDOR_MXL
 		case WFA_CAPA_OUI_TYPE:
 			elems->wfa_capab = pos + 4;
 			elems->wfa_capab_len = elen - 4;
 			break;
+#endif /* CONFIG_VENDOR_MXL */
 		case WFA_RSNE_OVERRIDE_OUI_TYPE:
 			elems->rsne_override = pos;
 			elems->rsne_override_len = elen;
diff --git a/src/common/wpa_common.c b/src/common/wpa_common.c
index cb0f05771..17ccfa118 100644
--- a/src/common/wpa_common.c
+++ b/src/common/wpa_common.c
@@ -2197,7 +2197,7 @@ int wpa_modify_wpa_ie_rsn(u8 *rsn_ie, size_t rsn_ie_len, u32 group_suite, u16 ca
 
 	if (rsn_ie_len >= 6 && rsn_ie[1] >= 4 &&
 		rsn_ie[1] == rsn_ie_len - 2 &&
-		WPA_GET_BE32(&rsn_ie[2]) == OSEN_IE_VENDOR_TYPE) {
+		WPA_GET_BE32(&rsn_ie[2]) == WPA_IE_VENDOR_TYPE) {
 		pos = rsn_ie + 6;
 		left = rsn_ie_len - 6;
 	} else {
diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index 700f4694c..a6249299b 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -2679,7 +2679,9 @@ struct hostapd_sta_add_params {
 	bool mld_link_sta;
 	s8 mld_link_id;
 	const u8 *mld_link_addr;
+#ifndef CONFIG_VENDOR_MXL
 	u16 eml_cap;
+#endif /* CONFIG_VENDOR_MXL */
 #ifdef CONFIG_VENDOR_MXL
 	int ssi_signal;
 	struct ieee80211_mgmt *last_assoc_req;
diff --git a/src/drivers/driver_common.c b/src/drivers/driver_common.c
index 85121033c..bfefa1316 100644
--- a/src/drivers/driver_common.c
+++ b/src/drivers/driver_common.c
@@ -106,8 +106,9 @@ const char * event_to_string(enum wpa_event_type event)
 	E2S(LINK_CH_SWITCH_STARTED);
 	E2S(TID_LINK_MAP);
 	E2S(LINK_RECONFIG);
+#ifndef CONFIG_VENDOR_MXL
 	E2S(MLD_INTERFACE_FREED);
-#ifdef CONFIG_VENDOR_MXL
+#else /* CONFIG_VENDOR_MXL */
 	E2S(LTQ_UNCONNECTED_STA_RX);
 	MXL_E2S;
 #endif /* CONFIG_VENDOR_MXL */
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index a61dc0dd2..466f78684 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -174,11 +174,18 @@ static void wpa_driver_nl80211_deinit(struct i802_bss *bss);
 static int wpa_driver_nl80211_set_mode_ibss(struct i802_bss *bss,
 					    struct hostapd_freq_params *freq);
 
+#ifndef CONFIG_VENDOR_MXL
 static int
 wpa_driver_nl80211_finish_drv_init(struct wpa_driver_nl80211_data *drv,
 				   const u8 *set_addr, int first,
 				   const char *driver_params,
 				   enum wpa_p2p_mode p2p_mode);
+#else /* CONFIG_VENDOR_MXL */
+static int
+wpa_driver_nl80211_finish_drv_init(struct wpa_driver_nl80211_data *drv,
+				   const u8 *set_addr, int first,
+				   const char *driver_params);
+#endif /* CONFIG_VENDOR_MXL */
 static int nl80211_send_frame_cmd(struct i802_bss *bss,
 				  unsigned int freq, unsigned int wait,
 				  const u8 *buf, size_t buf_len,
@@ -1248,8 +1255,12 @@ static int wpa_driver_nl80211_own_ifindex(struct wpa_driver_nl80211_data *drv,
 		nl80211_check_global(drv->global);
 		wpa_printf(MSG_DEBUG, "nl80211: Update ifindex for a removed "
 			   "interface");
+#ifndef CONFIG_VENDOR_MXL
 		if (wpa_driver_nl80211_finish_drv_init(drv, NULL, 0, NULL,
 						       WPA_P2P_MODE_WFD_R1) < 0)
+#else /* CONFIG_VENDOR_MXL */
+		if (wpa_driver_nl80211_finish_drv_init(drv, NULL, 0, NULL) < 0)
+#endif /* CONFIG_VENDOR_MXL */
 			return -1;
 		return 1;
 	}
@@ -2398,11 +2409,18 @@ wpa_driver_nl80211_drv_init_rfkill(struct wpa_driver_nl80211_data *drv)
 }
 
 
+#ifndef CONFIG_VENDOR_MXL
 static void * wpa_driver_nl80211_drv_init(void *ctx, const char *ifname,
 					  void *global_priv, int hostapd,
 					  const u8 *set_addr,
 					  const char *driver_params,
 					  enum wpa_p2p_mode p2p_mode)
+#else /* CONFIG_VENDOR_MXL */
+static void * wpa_driver_nl80211_drv_init(void *ctx, const char *ifname,
+					  void *global_priv, int hostapd,
+					  const u8 *set_addr,
+					  const char *driver_params)
+#endif /* CONFIG_VENDOR_MXL */
 {
 	struct wpa_driver_nl80211_data *drv;
 	struct i802_bss *bss;
@@ -2456,8 +2474,13 @@ static void * wpa_driver_nl80211_drv_init(void *ctx, const char *ifname,
 	bss->drv = drv;
 	bss->ctx = ctx;
 #ifdef CONFIG_VENDOR_MXL
-	for (i = 0; i < MAX_NUM_MLD_LINKS; i++)
-		bss->links[i].link_id = NL80211_DRV_LINK_ID_NA;
+	int link_id;
+	mxl_set_scan_timeout_default(bss, SCAN_TIMEOUT_DEFAULT);
+
+	for_each_link_default(bss->valid_links, link_id, NL80211_DRV_LINK_ID_NA)
+		if (link_id != NL80211_DRV_LINK_ID_NA) {
+			wpa_printf(MSG_DEBUG, "Failed to set Opensource ml_link_ID to NA");
+		}
 #endif /* CONFIG_VENDOR_MXL */
 
 	os_strlcpy(bss->ifname, ifname, sizeof(bss->ifname));
@@ -2467,8 +2490,12 @@ static void * wpa_driver_nl80211_drv_init(void *ctx, const char *ifname,
 	if (nl80211_init_bss(bss))
 		goto failed;
 
+#ifndef CONFIG_VENDOR_MXL
 	if (wpa_driver_nl80211_finish_drv_init(drv, set_addr, 1, driver_params,
 					       p2p_mode))
+#else /* CONFIG_VENDOR_MXL */
+	if (wpa_driver_nl80211_finish_drv_init(drv, set_addr, 1, driver_params))
+#endif /* CONFIG_VENDOR_MXL */
 		goto failed;
 
 	if (drv->capa.flags2 & WPA_DRIVER_FLAGS2_CONTROL_PORT_TX_STATUS) {
@@ -3198,11 +3225,18 @@ fail:
 #endif /* CONFIG_DRIVER_NL80211_QCA */
 
 
+#ifndef CONFIG_VENDOR_MXL
 static int
 wpa_driver_nl80211_finish_drv_init(struct wpa_driver_nl80211_data *drv,
 				   const u8 *set_addr, int first,
 				   const char *driver_params,
 				   enum wpa_p2p_mode p2p_mode)
+#else /* CONFIG_VENDOR_MXL */
+static int
+wpa_driver_nl80211_finish_drv_init(struct wpa_driver_nl80211_data *drv,
+				   const u8 *set_addr, int first,
+				   const char *driver_params)
+#endif /* CONFIG_VENDOR_MXL */
 {
 	struct i802_bss *bss = drv->first_bss;
 	int send_rfkill_event = 0;
@@ -3261,10 +3295,12 @@ wpa_driver_nl80211_finish_drv_init(struct wpa_driver_nl80211_data *drv,
 
 	wpa_driver_nl80211_drv_init_rfkill(drv);
 
+#ifndef CONFIG_VENDOR_MXL
 #ifdef CONFIG_DRIVER_NL80211_QCA
 	if (nlmode == NL80211_IFTYPE_P2P_GO)
 		nl80211_set_p2p_mode(bss, p2p_mode);
 #endif /* CONFIG_DRIVER_NL80211_QCA */
+#endif /* CONFIG_VENDOR_MXL */
 
 	if (!rfkill_is_blocked(drv->rfkill)) {
 		int ret = i802_set_iface_flags(bss, 1);
@@ -6008,6 +6044,7 @@ static int wpa_driver_nl80211_sta_add(void *priv,
 				goto fail;
 		}
 
+#ifndef CONFIG_VENDOR_MXL
 		/* Set EML capabilities of ML STA */
 		if (params->mld_link_addr && params->eml_cap) {
 			wpa_printf(MSG_DEBUG, "  * eml_cap =%u",
@@ -6016,6 +6053,7 @@ static int wpa_driver_nl80211_sta_add(void *priv,
 					params->eml_cap))
 				goto fail;
 		}
+#endif /* CONFIG_VENDOR_MXL */
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 		if (params->mxl_params.ml_sta_info && params->mxl_params.ml_sta_info->is_ml) {
@@ -6674,6 +6712,7 @@ static int wpa_driver_nl80211_hapd_send_eapol(
 		struct hostapd_data *hapd = (struct hostapd_data *)(bss->ctx);
 		u8 *vlan_data = NULL;
 		u16 proto = ETH_P_EAPOL;
+		int res = 0;
 		if ((flags & MXL_WPA_STA_MULTI_AP) && (hapd->conf->multi_ap == BACKHAUL_BSS) && hapd->conf->multi_ap_vlanid) {
 			proto = mxl_ieee802_11_add_vlan_hdr(&vlan_data, hapd->conf->multi_ap_vlanid, proto, data, &data_len);
 		}
@@ -9038,10 +9077,16 @@ static void *i802_init(struct hostapd_data *hapd,
 	int ifindex, br_ifindex = 0;
 	int br_added = 0;
 
+#ifndef CONFIG_VENDOR_MXL
 	bss = wpa_driver_nl80211_drv_init(hapd, params->ifname,
 					  params->global_priv, 1,
 					  params->bssid, params->driver_params,
 					  WPA_P2P_MODE_WFD_R1);
+#else /* CONFIG_VENDOR_MXL */
+	bss = wpa_driver_nl80211_drv_init(hapd, params->ifname,
+					  params->global_priv, 1,
+					  params->bssid, params->driver_params);
+#endif /* CONFIG_VENDOR_MXL */
 	if (bss == NULL)
 		return NULL;
 
@@ -11353,7 +11398,9 @@ static int driver_nl80211_link_remove(void *priv, enum wpa_driver_if_type type,
 			return ret;
 
 		/* Notify that the MLD interface is removed */
+#ifndef CONFIG_VENDOR_MXL
 		wpa_supplicant_event(bss->ctx, EVENT_MLD_INTERFACE_FREED, NULL);
+#endif /* CONFIG_VENDOR_MXL */
 	}
 
 	return 0;
@@ -12331,7 +12378,7 @@ static int nl80211_vendor_cmd(void *priv, unsigned int vendor_id,
 #ifdef CONFIG_VENDOR_MXL
 #ifdef CONFIG_DRIVER_NL80211_MXL
 	if (OUI_LTQ == vendor_id)
-		ret = send_and_recv_msgs(drv, msg, vendor_mxl_reply_handler, buf,
+		ret = send_and_recv(drv, drv->global->nl, msg, vendor_mxl_reply_handler, buf,
 					 NULL, NULL, NULL);
 	else
 #endif /* CONFIG_DRIVER_NL80211_MXL */
@@ -15515,7 +15562,7 @@ int nl80211_set_mbssid_num_vaps_in_group(void *priv, const u8 mbssid_num_vaps_in
 
 int nl80211_get_tx_power_20mhz(void *priv, struct mxl_vendor_20mhz_tx_power *tx_power)
 {
-	int ret;
+	int ret = 0;
 	errno_t err;
 	size_t tx_power_vendor_size = sizeof(struct mxl_vendor_20mhz_tx_power);
 	struct wpabuf *rsp_tx_power;
diff --git a/src/drivers/driver_nl80211_capa.c b/src/drivers/driver_nl80211_capa.c
index 9cdec7af6..27ac90202 100644
--- a/src/drivers/driver_nl80211_capa.c
+++ b/src/drivers/driver_nl80211_capa.c
@@ -1525,7 +1525,7 @@ static void _ltq_nl80211_get_prop_phy_cap(struct wpa_driver_nl80211_data *drv, s
 		nlmsg_free(msg);
 		return;
 	}
-	ret = send_and_recv_msgs(drv, msg, _get_prop_phy_cap_reply_handler, cap, NULL, NULL, NULL);
+	ret = send_and_recv(drv, drv->global->nl, msg, _get_prop_phy_cap_reply_handler, cap, NULL, NULL, NULL);
 	if (ret) {
 		wpa_printf(MSG_ERROR, "nl80211: sending/receiving GET_PROP_PHY_CAP failed: %i (%s)", ret, strerror(-ret));
 		return;
diff --git a/src/drivers/driver_nl80211_event.c b/src/drivers/driver_nl80211_event.c
index 3eeb47d7b..ee346950b 100644
--- a/src/drivers/driver_nl80211_event.c
+++ b/src/drivers/driver_nl80211_event.c
@@ -13,6 +13,7 @@
 
 #include "utils/common.h"
 #include "utils/eloop.h"
+#include "ap/hostapd.h"
 #include "common/qca-vendor.h"
 #include "common/qca-vendor-attr.h"
 #include "common/brcm_vendor.h"
@@ -2610,7 +2611,7 @@ static void nl80211_process_radar_event(struct i802_bss *bss,
 		break;
 #if defined(CONFIG_VENDOR_MXL) && !defined(CONFIG_VENDOR_MXL_BUILD_FOR_SUPPLICANT)
 	case NL80211_RADAR_CAC_PAUSED:
-		mxl_hostapd_event_dfs_cac_paused(bss->ctx, data->dfs_event);
+		mxl_hostapd_event_dfs_cac_paused(bss->ctx, &data->dfs_event);
 		break;
 #endif /* defined(CONFIG_VENDOR_MXL) && !defined(CONFIG_VENDOR_MXL_BUILD_FOR_SUPPLICANT) */
 	default:
@@ -4021,7 +4022,7 @@ static void nl80211_control_port_frame(struct i802_bss *bss, struct nlattr **tb)
 		break;
 	case ETH_P_PAE:
 #ifdef CONFIG_VENDOR_MXL
-		mxl_drv_event_eapol_rx2(drv->ctx, src_addr,
+		mxl_drv_event_eapol_rx2(bss->drv->ctx, src_addr,
 					nla_data(tb[NL80211_ATTR_FRAME]),
 					nla_len(tb[NL80211_ATTR_FRAME]),
 					encrypted, link_id,
diff --git a/src/drivers/mxl_driver_nl80211.c b/src/drivers/mxl_driver_nl80211.c
index 0e8855cf0..01a7fa2d2 100644
--- a/src/drivers/mxl_driver_nl80211.c
+++ b/src/drivers/mxl_driver_nl80211.c
@@ -45,7 +45,7 @@ int mxl_nl80211_abort_scan(struct i802_bss *bss)
 
 	wpa_printf(MSG_DEBUG, "nl80211: Abort scan");
 	msg = nl80211_cmd_msg(bss, 0, NL80211_CMD_ABORT_SCAN);
-	ret = send_and_recv_msgs(drv, msg, NULL, NULL, NULL, NULL, NULL);
+	ret = send_and_recv(drv, drv->global->nl, msg, NULL, NULL, NULL, NULL, NULL);
 	if (ret) {
 		wpa_printf(MSG_DEBUG, "nl80211: Abort scan failed: ret=%d (%s)",
 			   ret, strerror(-ret));
@@ -87,7 +87,7 @@ int mxl_nl80211_set_antenna(void *priv, u32 tx_ant, u32 rx_ant)
 		return -1;
 	}
 
-	ret = send_and_recv_msgs(drv, msg, NULL, NULL, NULL, NULL, NULL);
+	ret = send_and_recv(drv, drv->global->nl, msg, NULL, NULL, NULL, NULL, NULL);
 	if (ret != 0) {
 		wpa_printf(MSG_DEBUG, "nl80211: Failed to set antenna mask: "
 			"%d (%s)", ret, strerror(-ret));
@@ -119,7 +119,7 @@ int mxl_nl80211_set_channel_dfs_state(void *priv, enum nl80211_dfs_state dfs_sta
 		return -1;
 	}
 
-	ret = send_and_recv_msgs(drv, msg, NULL, NULL, NULL, NULL, NULL);
+	ret = send_and_recv(drv, drv->global->nl, msg, NULL, NULL, NULL, NULL, NULL);
 	if (ret == 0)
 		return 0;
 	wpa_printf(MSG_DEBUG, "nl80211: Failed to set channel DFS state: "
diff --git a/wpa_supplicant/Makefile b/wpa_supplicant/Makefile
index 6cacbaeac..bd8b46452 100644
--- a/wpa_supplicant/Makefile
+++ b/wpa_supplicant/Makefile
@@ -123,7 +123,6 @@ OBJS_c = wpa_cli.o ../src/common/wpa_ctrl.o
 OBJS_c += ../src/utils/wpa_debug.o
 OBJS_c += ../src/utils/common.o
 OBJS_c += ../src/common/cli.o
-OBJS += wmm_ac.o
 OBJS += ../src/common/wpa_ctrl.o
 
 ifndef CONFIG_OS
diff --git a/wpa_supplicant/bgscan_simple.c b/wpa_supplicant/bgscan_simple.c
index d9aaa634f..18d4f868e 100644
--- a/wpa_supplicant/bgscan_simple.c
+++ b/wpa_supplicant/bgscan_simple.c
@@ -58,6 +58,7 @@ static bool bgscan_simple_btm_query(struct wpa_supplicant *wpa_s,
 	wpa_printf(MSG_DEBUG,
 		   "bgscan simple: Send BSS transition management query %d/%d",
 		   mod, data->use_btm_query);
+#ifndef CONFIG_VENDOR_MXL
 	if (wnm_send_bss_transition_mgmt_query(
 		    wpa_s, WNM_TRANSITION_REASON_BETTER_AP_FOUND, NULL, 0)) {
 		wpa_printf(MSG_DEBUG,
@@ -65,6 +66,7 @@ static bool bgscan_simple_btm_query(struct wpa_supplicant *wpa_s,
 		/* Fall through and do regular scan */
 		return false;
 	}
+#endif /* CONFIG_VENDOR_MXL */
 
 	/* Start a new timeout for the next one. We don't have scan callback to
 	 * otherwise trigger future progress when using BTM path. */
diff --git a/wpa_supplicant/dbus/dbus_new.c b/wpa_supplicant/dbus/dbus_new.c
index 7893f3561..873b8b34d 100644
--- a/wpa_supplicant/dbus/dbus_new.c
+++ b/wpa_supplicant/dbus/dbus_new.c
@@ -13,6 +13,7 @@
 #include "common.h"
 #include "common/ieee802_11_defs.h"
 #include "wps/wps.h"
+#include "ap/hostapd.h"
 #include "ap/sta_info.h"
 #include "../config.h"
 #include "../wpa_supplicant_i.h"
diff --git a/wpa_supplicant/driver_i.h b/wpa_supplicant/driver_i.h
index 406ebd202..c03257188 100644
--- a/wpa_supplicant/driver_i.h
+++ b/wpa_supplicant/driver_i.h
@@ -22,13 +22,13 @@ static inline void * wpa_drv_init(struct wpa_supplicant *wpa_s,
 #endif /* CONFIG_VENDOR_MXL */
 {
 	if (wpa_s->driver->init2) {
+#ifndef CONFIG_VENDOR_MXL
 		enum wpa_p2p_mode p2p_mode = WPA_P2P_MODE_WFD_R1;
 
 #ifdef CONFIG_P2P
 		p2p_mode = wpa_s->p2p_mode;
 #endif /* CONFIG_P2P */
 
-#ifndef CONFIG_VENDOR_MXL
 		return wpa_s->driver->init2(wpa_s, ifname,
 					    wpa_s->global_drv_priv,
 					    p2p_mode);
diff --git a/wpa_supplicant/mesh.c b/wpa_supplicant/mesh.c
index 869f0b39f..4882d65a3 100644
--- a/wpa_supplicant/mesh.c
+++ b/wpa_supplicant/mesh.c
@@ -14,8 +14,8 @@
 #include "common/ieee802_11_defs.h"
 #include "common/wpa_ctrl.h"
 #include "common/hw_features_common.h"
-#include "ap/sta_info.h"
 #include "ap/hostapd.h"
+#include "ap/sta_info.h"
 #include "ap/ieee802_11.h"
 #include "config_ssid.h"
 #include "config.h"
@@ -215,6 +215,7 @@ static int wpas_mesh_update_freq_params(struct wpa_supplicant *wpa_s)
 	if (ifmsh->current_mode)
 		he_capab = &ifmsh->current_mode->he_capab[IEEE80211_MODE_MESH];
 
+#ifndef CONFIG_VENDOR_MXL
 	if (hostapd_set_freq_params(
 		    &params->freq,
 		    ifmsh->conf->hw_mode,
@@ -232,6 +233,25 @@ static int wpas_mesh_update_freq_params(struct wpa_supplicant *wpa_s)
 		    hostapd_get_oper_centr_freq_seg1_idx(ifmsh->conf),
 		    ifmsh->conf->vht_capab,
 		    he_capab, NULL, 0)) {
+#else /* CONFIG_VENDOR_MXL */
+	if (hostapd_set_freq_params(
+		    &params->freq,
+		    ifmsh->conf->hw_mode,
+		    ifmsh->freq,
+		    ifmsh->conf->channel,
+		    ifmsh->conf->enable_edmg,
+		    ifmsh->conf->edmg_channel,
+		    ifmsh->conf->ieee80211n,
+		    ifmsh->conf->ieee80211ac,
+		    ifmsh->conf->ieee80211ax,
+		    ifmsh->conf->ieee80211be,
+		    ifmsh->conf->secondary_channel,
+		    hostapd_get_oper_chwidth(ifmsh->conf),
+		    hostapd_get_oper_centr_freq_seg0_idx(ifmsh->conf),
+		    hostapd_get_oper_centr_freq_seg1_idx(ifmsh->conf),
+		    ifmsh->conf->vht_capab,
+		    he_capab, NULL)) {
+#endif /* CONFIG_VENDOR_MXL */
 		wpa_printf(MSG_ERROR, "Error updating mesh frequency params");
 		wpa_supplicant_mesh_deinit(wpa_s, true);
 		return -1;
diff --git a/wpa_supplicant/mxl_supplicant.c b/wpa_supplicant/mxl_supplicant.c
index c2c849d7c..4a6623c5f 100644
--- a/wpa_supplicant/mxl_supplicant.c
+++ b/wpa_supplicant/mxl_supplicant.c
@@ -17,15 +17,16 @@
 #include <stdlib.h>
 #include <string.h>
 #include "utils/common.h"
+#include "ap/hostapd.h"
 #include "common/ieee802_11_defs.h"
 #include "common/ieee802_11_common.h"
 #include "common/mxl_ieee802_11_common.h"
 #include "config.h"
 #include "ap/mxl_hostapd.h"
 #include "mxl_supplicant_config.h"
+#include "wpa_supplicant_i.h"
 #include "bss.h"
 #include "driver_i.h"
-#include "wpa_supplicant_i.h"
 
 int mxl_supplicant_drv_get_vendor_data(struct wpa_supplicant *wpa_s,
 			int vendor_id, int subcmd,
diff --git a/wpa_supplicant/p2p_supplicant.c b/wpa_supplicant/p2p_supplicant.c
index 6d2551b89..d1d42bd99 100644
--- a/wpa_supplicant/p2p_supplicant.c
+++ b/wpa_supplicant/p2p_supplicant.c
@@ -2255,11 +2255,18 @@ static void p2p_go_configured(void *ctx, void *data)
 	    params->akmp == WPA_KEY_MGMT_SAE) {
 		struct hostapd_data *hapd = wpa_s->ap_iface->bss[0];
 
+#ifndef CONFIG_VENDOR_MXL
 		wpa_auth_pmksa_add_sae(hapd->wpa_auth,
 				       params->peer_device_addr,
 				       params->pmk, params->pmk_len,
 				       params->pmkid, WPA_KEY_MGMT_SAE,
 				       false);
+#else /* CONFIG_VENDOR_MXL */
+		wpa_auth_pmksa_add_sae(hapd->wpa_auth,
+				       params->peer_device_addr,
+				       params->pmk, params->pmk_len,
+				       params->pmkid, WPA_KEY_MGMT_SAE);
+#endif /* CONFIG_VENDOR_MXL */
 		hostapd_add_pmkid(hapd, params->peer_device_addr,
 				  params->pmk, params->pmk_len,
 				  params->pmkid, WPA_KEY_MGMT_SAE);
@@ -2911,11 +2918,18 @@ static void wpas_set_go_security_config(void *ctx,
 	if (params->akmp == WPA_KEY_MGMT_SAE) {
 		wpa_printf(MSG_DEBUG, "P2P: Adding PMK for peer: " MACSTR,
 			   MAC2STR(params->peer_device_addr));
+#ifndef CONFIG_VENDOR_MXL
 		wpa_auth_pmksa_add_sae(hapd->wpa_auth,
 				       params->peer_device_addr,
 				       params->pmk, params->pmk_len,
 				       params->pmkid, WPA_KEY_MGMT_SAE,
 				       false);
+#else /* CONFIG_VENDOR_MXL */
+		wpa_auth_pmksa_add_sae(hapd->wpa_auth,
+				       params->peer_device_addr,
+				       params->pmk, params->pmk_len,
+				       params->pmkid, WPA_KEY_MGMT_SAE);
+#endif /* CONFIG_VENDOR_MXL */
 		hostapd_add_pmkid(hapd, params->peer_device_addr,
 				  params->pmk, params->pmk_len,
 				  params->pmkid, WPA_KEY_MGMT_SAE);
diff --git a/wpa_supplicant/wpa_supplicant.c b/wpa_supplicant/wpa_supplicant.c
index 0b1424f91..04b98dd40 100644
--- a/wpa_supplicant/wpa_supplicant.c
+++ b/wpa_supplicant/wpa_supplicant.c
@@ -3512,7 +3512,11 @@ skip_80mhz:
 				    freq->sec_channel_offset,
 				    chwidth, seg0, seg1, vht_caps,
 				    &mode->he_capab[ieee80211_mode],
+#ifndef CONFIG_VENDOR_MXL
 				    &mode->eht_capab[ieee80211_mode], 0) != 0)
+#else /* CONFIG_VENDOR_MXL */
+				    &mode->eht_capab[ieee80211_mode]) != 0)
+#endif /* CONFIG_VENDOR_MXL */
 		return false;
 
 	*freq = vht_freq;
-- 
2.43.0

